// lucid-ui/dist/esm/components/ExpanderPanel/ExpanderPanel.stories.js
import _ from 'lodash';
import React from 'react';
import ExpanderPanel from './ExpanderPanel';
export default {
    title: 'Layout/ExpanderPanel',
    component: ExpanderPanel,
    parameters: {
        docs: {
            description: {
                component: ExpanderPanel.peek.description,
            },
        },
    },
    args: ExpanderPanel.defaultProps,
};
/* Basic */
export const Basic = (args) => {
    return (React.createElement(ExpanderPanel, { ...args },
        React.createElement(ExpanderPanel.Header, null, "Show More"),
        _.times(100, (n) => (React.createElement("div", { key: n }, _.repeat('-', 75 * Math.sin(n / 5)))))));
};
/* No Padding */
export const NoPadding = (args) => {
    return (React.createElement(ExpanderPanel, { ...args, hasPadding: false },
        React.createElement(ExpanderPanel.Header, null, "Show More"),
        _.times(100, (n) => (React.createElement("div", { key: n }, _.repeat('-', 75 * Math.sin(n / 5)))))));
};
/* Basic With On Rest Callback */
export const BasicWithOnRestCallback = (args) => {
    const onRest = () => {
        alert('A big ball of wibbly wobbly, timey wimey stuff');
    };
    return (React.createElement(ExpanderPanel, { ...args, onRest: onRest },
        React.createElement(ExpanderPanel.Header, null, "Show More"),
        _.times(100, (n) => (React.createElement("div", { key: n }, _.repeat('-', 75 * Math.sin(n / 5)))))));
};
