// lucid-ui/dist/cjs/constants/charts.js
"use strict";
// NOTE: these needs to stay in sync with variables.less
Object.defineProperty(exports, "__esModule", { value: true });
exports.PALETTE_MONOCHROME_GOOD_3 = exports.PALETTE_30 = exports.PALETTE_7 = exports.PALETTE_6 = exports.COLOR_NEUTRAL = exports.COLOR_BAD_DARKEST = exports.COLOR_BAD_DARK = exports.COLOR_BAD_LIGHT = exports.COLOR_BAD_LIGHTEST = exports.COLOR_BAD = exports.COLOR_GOOD_DARKEST = exports.COLOR_GOOD_DARK = exports.COLOR_GOOD_LIGHT = exports.COLOR_GOOD_LIGHTEST = exports.COLOR_GOOD = exports.COLOR_6_DARKEST = exports.COLOR_6_DARK = exports.COLOR_6_LIGHT = exports.COLOR_6_LIGHTEST = exports.COLOR_6 = exports.COLOR_5_DARKEST = exports.COLOR_5_DARK = exports.COLOR_5_LIGHT = exports.COLOR_5_LIGHTEST = exports.COLOR_5 = exports.COLOR_4_DARKEST = exports.COLOR_4_DARK = exports.COLOR_4_LIGHT = exports.COLOR_4_LIGHTEST = exports.COLOR_4 = exports.COLOR_3_DARKEST = exports.COLOR_3_DARK = exports.COLOR_3_LIGHT = exports.COLOR_3_LIGHTEST = exports.COLOR_3 = exports.COLOR_2_DARKEST = exports.COLOR_2_DARK = exports.COLOR_2_LIGHT = exports.COLOR_2_LIGHTEST = exports.COLOR_2 = exports.COLOR_1_DARKEST = exports.COLOR_1_DARK = exports.COLOR_1_LIGHT = exports.COLOR_1_LIGHTEST = exports.COLOR_1 = exports.COLOR_0_DARKEST = exports.COLOR_0_DARK = exports.COLOR_0_LIGHT = exports.COLOR_0_LIGHTEST = exports.COLOR_0 = void 0;
exports.PALETTE_MONOCHROME_6_5 = exports.PALETTE_MONOCHROME_5_5 = exports.PALETTE_MONOCHROME_4_5 = exports.PALETTE_MONOCHROME_3_5 = exports.PALETTE_MONOCHROME_2_5 = exports.PALETTE_MONOCHROME_1_5 = exports.PALETTE_MONOCHROME_0_5 = exports.PALETTE_MONOCHROME_BAD_5 = exports.PALETTE_MONOCHROME_BAD_3 = exports.PALETTE_MONOCHROME_GOOD_5 = void 0;
// Each of these map to a less variable that applies the actual color
exports.COLOR_0 = 'color-chart-0';
exports.COLOR_0_LIGHTEST = 'color-chart-0-lightest';
exports.COLOR_0_LIGHT = 'color-chart-0-light';
exports.COLOR_0_DARK = 'color-chart-0-dark';
exports.COLOR_0_DARKEST = 'color-chart-0-darkest';
exports.COLOR_1 = 'color-chart-1';
exports.COLOR_1_LIGHTEST = 'color-chart-1-lightest';
exports.COLOR_1_LIGHT = 'color-chart-1-light';
exports.COLOR_1_DARK = 'color-chart-1-dark';
exports.COLOR_1_DARKEST = 'color-chart-1-darkest';
exports.COLOR_2 = 'color-chart-2';
exports.COLOR_2_LIGHTEST = 'color-chart-2-lightest';
exports.COLOR_2_LIGHT = 'color-chart-2-light';
exports.COLOR_2_DARK = 'color-chart-2-dark';
exports.COLOR_2_DARKEST = 'color-chart-2-darkest';
exports.COLOR_3 = 'color-chart-3';
exports.COLOR_3_LIGHTEST = 'color-chart-3-lightest';
exports.COLOR_3_LIGHT = 'color-chart-3-light';
exports.COLOR_3_DARK = 'color-chart-3-dark';
exports.COLOR_3_DARKEST = 'color-chart-3-darkest';
exports.COLOR_4 = 'color-chart-4';
exports.COLOR_4_LIGHTEST = 'color-chart-4-lightest';
exports.COLOR_4_LIGHT = 'color-chart-4-light';
exports.COLOR_4_DARK = 'color-chart-4-dark';
exports.COLOR_4_DARKEST = 'color-chart-4-darkest';
exports.COLOR_5 = 'color-chart-5';
exports.COLOR_5_LIGHTEST = 'color-chart-5-lightest';
exports.COLOR_5_LIGHT = 'color-chart-5-light';
exports.COLOR_5_DARK = 'color-chart-5-dark';
exports.COLOR_5_DARKEST = 'color-chart-5-darkest';
exports.COLOR_6 = 'color-chart-6';
exports.COLOR_6_LIGHTEST = 'color-chart-6-lightest';
exports.COLOR_6_LIGHT = 'color-chart-6-light';
exports.COLOR_6_DARK = 'color-chart-6-dark';
exports.COLOR_6_DARKEST = 'color-chart-6-darkest';
exports.COLOR_GOOD = 'color-chart-good';
exports.COLOR_GOOD_LIGHTEST = 'color-chart-good-lightest';
exports.COLOR_GOOD_LIGHT = 'color-chart-good-light';
exports.COLOR_GOOD_DARK = 'color-chart-good-dark';
exports.COLOR_GOOD_DARKEST = 'color-chart-good-darkest';
exports.COLOR_BAD = 'color-chart-bad';
exports.COLOR_BAD_LIGHTEST = 'color-chart-bad-lightest';
exports.COLOR_BAD_LIGHT = 'color-chart-bad-light';
exports.COLOR_BAD_DARK = 'color-chart-bad-dark';
exports.COLOR_BAD_DARKEST = 'color-chart-bad-darkest';
exports.COLOR_NEUTRAL = 'color-chart-neutral';
exports.PALETTE_6 = [
    exports.COLOR_3,
    exports.COLOR_0_DARKEST,
    exports.COLOR_2,
    exports.COLOR_4_LIGHTEST,
    exports.COLOR_5,
    exports.COLOR_1_LIGHT,
];
// In order to keep the naming scheme correct I'm introducing a new default
// palette that has 7 items while not breaking existing code that uses a 6
// color palette.
exports.PALETTE_7 = [
    exports.COLOR_3,
    exports.COLOR_0_DARKEST,
    exports.COLOR_2,
    exports.COLOR_4_LIGHTEST,
    exports.COLOR_5,
    exports.COLOR_1_LIGHT,
    exports.COLOR_6,
];
exports.PALETTE_30 = [
    exports.COLOR_0_LIGHTEST,
    exports.COLOR_0_LIGHT,
    exports.COLOR_0,
    exports.COLOR_0_DARK,
    exports.COLOR_0_DARKEST,
    exports.COLOR_1_LIGHTEST,
    exports.COLOR_1_LIGHT,
    exports.COLOR_1,
    exports.COLOR_1_DARK,
    exports.COLOR_1_DARKEST,
    exports.COLOR_2_LIGHTEST,
    exports.COLOR_2_LIGHT,
    exports.COLOR_2,
    exports.COLOR_2_DARK,
    exports.COLOR_2_DARKEST,
    exports.COLOR_3_LIGHTEST,
    exports.COLOR_3_LIGHT,
    exports.COLOR_3,
    exports.COLOR_3_DARK,
    exports.COLOR_3_DARKEST,
    exports.COLOR_4_LIGHTEST,
    exports.COLOR_4_LIGHT,
    exports.COLOR_4,
    exports.COLOR_4_DARK,
    exports.COLOR_4_DARKEST,
    exports.COLOR_5_LIGHTEST,
    exports.COLOR_5_LIGHT,
    exports.COLOR_5,
    exports.COLOR_5_DARK,
    exports.COLOR_5_DARKEST,
];
exports.PALETTE_MONOCHROME_GOOD_3 = [
    exports.COLOR_GOOD_LIGHT,
    exports.COLOR_GOOD,
    exports.COLOR_GOOD_DARK,
];
exports.PALETTE_MONOCHROME_GOOD_5 = [
    exports.COLOR_GOOD_LIGHTEST,
    exports.COLOR_GOOD_LIGHT,
    exports.COLOR_GOOD,
    exports.COLOR_GOOD_DARK,
    exports.COLOR_GOOD_DARKEST,
];
exports.PALETTE_MONOCHROME_BAD_3 = [
    exports.COLOR_BAD_LIGHT,
    exports.COLOR_BAD,
    exports.COLOR_BAD_DARK,
];
exports.PALETTE_MONOCHROME_BAD_5 = [
    exports.COLOR_BAD_LIGHTEST,
    exports.COLOR_BAD_LIGHT,
    exports.COLOR_BAD,
    exports.COLOR_BAD_DARK,
    exports.COLOR_BAD_DARKEST,
];
exports.PALETTE_MONOCHROME_0_5 = [
    exports.COLOR_0_LIGHTEST,
    exports.COLOR_0_LIGHT,
    exports.COLOR_0,
    exports.COLOR_0_DARK,
    exports.COLOR_0_DARKEST,
];
exports.PALETTE_MONOCHROME_1_5 = [
    exports.COLOR_1_LIGHTEST,
    exports.COLOR_1_LIGHT,
    exports.COLOR_1,
    exports.COLOR_1_DARK,
    exports.COLOR_1_DARKEST,
];
exports.PALETTE_MONOCHROME_2_5 = [
    exports.COLOR_2_LIGHTEST,
    exports.COLOR_2_LIGHT,
    exports.COLOR_2,
    exports.COLOR_2_DARK,
    exports.COLOR_2_DARKEST,
];
exports.PALETTE_MONOCHROME_3_5 = [
    exports.COLOR_3_LIGHTEST,
    exports.COLOR_3_LIGHT,
    exports.COLOR_3,
    exports.COLOR_3_DARK,
    exports.COLOR_3_DARKEST,
];
exports.PALETTE_MONOCHROME_4_5 = [
    exports.COLOR_4_LIGHTEST,
    exports.COLOR_4_LIGHT,
    exports.COLOR_4,
    exports.COLOR_4_DARK,
    exports.COLOR_4_DARKEST,
];
exports.PALETTE_MONOCHROME_5_5 = [
    exports.COLOR_5_LIGHTEST,
    exports.COLOR_5_LIGHT,
    exports.COLOR_5,
    exports.COLOR_5_DARK,
    exports.COLOR_5_DARKEST,
];
exports.PALETTE_MONOCHROME_6_5 = [
    exports.COLOR_6_LIGHTEST,
    exports.COLOR_6_LIGHT,
    exports.COLOR_6,
    exports.COLOR_6_DARK,
    exports.COLOR_6_DARKEST,
];
