// lucid-ui/dist/cjs/components/VerticalTabs/VerticalTabs.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VerticalTabsDumb = void 0;
var lodash_1 = __importStar(require("lodash"));
var react_1 = __importDefault(require("react"));
var prop_types_1 = __importDefault(require("prop-types"));
var style_helpers_1 = require("../../util/style-helpers");
var component_types_1 = require("../../util/component-types");
var state_management_1 = require("../../util/state-management");
var reducers = __importStar(require("./VerticalTabs.reducers"));
var VerticalListMenu_1 = require("../VerticalListMenu/VerticalListMenu");
var cx = style_helpers_1.lucidClassNames.bind('&-VerticalTabs');
var string = prop_types_1.default.string, number = prop_types_1.default.number, bool = prop_types_1.default.bool, func = prop_types_1.default.func;
var Tab = function (_props) { return null; };
Tab.displayName = 'VerticalTabs.Tab';
Tab.peek = {
    description: "Content that will be rendered in a tab. Be sure to nest a `Title` inside each `Tab` or provide it as a prop.",
};
Tab.propName = 'Tab';
Tab.propTypes = {
    /**
        Determines if the Tab is selected.
    */
    isSelected: bool,
};
/** Vertical Tabs Title Child Component */
var Title = function (_props) { return null; };
Title.displayName = 'VerticalTabs.Title';
Title.peek = {
    description: "A `Title` can be provided as a child or prop to a `Tab`.",
};
Title.propName = 'Title';
/** Default props for the VerticalTabs component */
var defaultProps = {
    selectedIndex: 0,
    onSelect: lodash_1.default.noop,
};
var VerticalTabs = /** @class */ (function (_super) {
    __extends(VerticalTabs, _super);
    function VerticalTabs() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    VerticalTabs.prototype.render = function () {
        var _a = this.props, className = _a.className, onSelect = _a.onSelect, selectedIndex = _a.selectedIndex, passThroughs = __rest(_a, ["className", "onSelect", "selectedIndex"]);
        // Grab props array from each Tab
        var tabChildProps = lodash_1.default.map((0, component_types_1.findTypes)(this.props, VerticalTabs.Tab), 'props');
        var selectedIndexFromChildren = lodash_1.default.findLastIndex(tabChildProps, {
            isSelected: true,
        });
        var actualSelectedIndex = selectedIndexFromChildren !== -1
            ? selectedIndexFromChildren
            : selectedIndex;
        return (react_1.default.createElement("div", __assign({}, (0, lodash_1.omit)(passThroughs, [
            'className',
            'selectedIndex',
            'onSelect',
            'initialState',
            'callbackId',
        ]), { className: cx('&', className) }),
            react_1.default.createElement(VerticalListMenu_1.VerticalListMenuDumb, { selectedIndices: [actualSelectedIndex], onSelect: onSelect }, lodash_1.default.map(tabChildProps, function (tabChildProp, index) { return (react_1.default.createElement(VerticalListMenu_1.VerticalListMenuDumb.Item, { className: cx('&-Tab', {
                    '&-Tab-is-active': actualSelectedIndex === index,
                }), key: index },
                react_1.default.createElement("span", { className: cx('&-Tab-content') }, lodash_1.default.get((0, component_types_1.getFirst)(tabChildProp, VerticalTabs.Title), 'props.children', '')))); })),
            react_1.default.createElement("div", { className: cx('&-content') }, lodash_1.default.get(tabChildProps, [actualSelectedIndex, 'children']))));
    };
    VerticalTabs.displayName = 'VerticalTabs';
    VerticalTabs.propTypes = {
        /**
            Class names that are appended to the defaults.
        */
        className: string,
        /**
            Indicates which of the \`VerticalTabs.Tab\` children is currently
            selected. The index of the last \`VerticalTabs.Tab\` child with
            \`isSelected\` equal to \`true\` takes precedence over this prop.
        */
        selectedIndex: number,
        /**
            Callback for when the user clicks a tab. Called with the index of the tab
            that was clicked.  Signature: \`(index, { event, props}) => {}\`
        */
        onSelect: func,
    };
    VerticalTabs.defaultProps = defaultProps;
    VerticalTabs.reducers = reducers;
    VerticalTabs.Tab = Tab;
    VerticalTabs.Title = Title;
    VerticalTabs.peek = {
        description: "`VerticalTabs` provides vertically tabbed navigation. It has a flexible interface that allows tab content to be passed as regular React children or through props.",
        categories: ['navigation'],
        madeFrom: ['VerticalListMenu'],
    };
    return VerticalTabs;
}(react_1.default.Component));
exports.VerticalTabsDumb = VerticalTabs;
exports.default = (0, state_management_1.buildModernHybridComponent)(VerticalTabs, { reducers: reducers });
