// lucid-ui/dist/cjs/components/Point/Point.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BigWithStroke = exports.Basic = void 0;
var lodash_1 = __importDefault(require("lodash"));
var react_1 = __importDefault(require("react"));
var create_react_class_1 = __importDefault(require("create-react-class"));
var Point_1 = __importDefault(require("./Point"));
var chartConstants = __importStar(require("../../constants/charts"));
exports.default = {
    title: 'Visualizations/Point',
    component: Point_1.default,
    parameters: {
        docs: {
            description: {
                component: Point_1.default.peek.description,
            },
        },
    },
};
/* Basic */
var Basic = function () {
    var svgProps = {
        width: 20,
        height: 20,
    };
    var pointProps = {
        x: 10,
        y: 10,
    };
    var Component = (0, create_react_class_1.default)({
        render: function () {
            return (react_1.default.createElement("div", null, lodash_1.default.map(chartConstants.PALETTE_7, function (color, i) { return (react_1.default.createElement("svg", __assign({ key: i }, svgProps),
                react_1.default.createElement(Point_1.default, __assign({}, pointProps, { kind: i, color: color })))); })));
        },
    });
    return react_1.default.createElement(Component, null);
};
exports.Basic = Basic;
/* Big With Stroke */
var BigWithStroke = function () {
    var svgProps = {
        width: 40,
        height: 40,
        style: {
            backgroundColor: 'black',
        },
    };
    var pointProps = {
        x: 20,
        y: 20,
        scale: 2,
        hasStroke: true,
    };
    var Component = (0, create_react_class_1.default)({
        render: function () {
            return (react_1.default.createElement("div", null, lodash_1.default.map(chartConstants.PALETTE_7, function (color, i) { return (react_1.default.createElement("svg", __assign({ key: i }, svgProps),
                react_1.default.createElement(Point_1.default, __assign({}, pointProps, { kind: i, color: color })))); })));
        },
    });
    return react_1.default.createElement(Component, null);
};
exports.BigWithStroke = BigWithStroke;
exports.BigWithStroke.storyName = 'BigWithStroke';
