// lucid-ui/dist/cjs/components/ExampleComponent/ExampleComponent.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PropExample = exports.Basic = void 0;
var react_1 = __importDefault(require("react"));
var ExampleComponent_1 = require("./ExampleComponent");
exports.default = {
    title: 'Documentation/ExampleComponent',
    component: ExampleComponent_1.ExampleComponent,
    parameters: {
        docs: {
            description: {
                component: ExampleComponent_1.ExampleComponent.description,
            },
        },
    },
    argTypes: {
        children: {
            control: false,
        },
    },
};
var Basic = function (args) {
    return (react_1.default.createElement(ExampleComponent_1.ExampleComponent, __assign({}, args), "Feel free to play with this example component to see how to create your own."));
};
exports.Basic = Basic;
var PropExample = function (args) {
    return (react_1.default.createElement(ExampleComponent_1.ExampleComponent, __assign({}, args, { isX: true }), "Be sure to show consumers of your component examples of how each prop is implemented."));
};
exports.PropExample = PropExample;
