// lucid-ui/dist/cjs/components/NotchedTag/NotchedTag.stories.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Basic = void 0;
var react_1 = __importDefault(require("react"));
var create_react_class_1 = __importDefault(require("create-react-class"));
var NotchedTag_1 = __importDefault(require("./NotchedTag"));
exports.default = {
    title: 'Visual Design/NotchedTag',
    component: NotchedTag_1.default,
    parameters: {
        docs: {
            description: {
                component: NotchedTag_1.default.peek.description,
            },
        },
    },
};
/* Basic */
var Basic = function () {
    var style = { display: 'inline-block', marginRight: '10px' };
    var Component = (0, create_react_class_1.default)({
        render: function () {
            return (react_1.default.createElement("div", null,
                react_1.default.createElement("div", null,
                    react_1.default.createElement(NotchedTag_1.default, { style: style }, "ADV"),
                    react_1.default.createElement(NotchedTag_1.default, { size: 'small', style: style }, "ADV"),
                    react_1.default.createElement(NotchedTag_1.default, { type: 'filled', style: style }, "ADV"),
                    react_1.default.createElement(NotchedTag_1.default, { size: 'small', type: 'filled', style: style }, "ADV")),
                react_1.default.createElement("div", null,
                    react_1.default.createElement(NotchedTag_1.default, { tagStyle: 'style-two', style: style }, "IO"),
                    react_1.default.createElement(NotchedTag_1.default, { size: 'small', tagStyle: 'style-two', style: style }, "IO"),
                    react_1.default.createElement(NotchedTag_1.default, { tagStyle: 'style-two', type: 'filled', style: style }, "IO"),
                    react_1.default.createElement(NotchedTag_1.default, { size: 'small', tagStyle: 'style-two', type: 'filled', style: style }, "IO")),
                react_1.default.createElement("div", null,
                    react_1.default.createElement(NotchedTag_1.default, { tagStyle: 'style-three', style: style }, "LI"),
                    react_1.default.createElement(NotchedTag_1.default, { size: 'small', tagStyle: 'style-three', style: style }, "LI"),
                    react_1.default.createElement(NotchedTag_1.default, { tagStyle: 'style-three', type: 'filled', style: style }, "LI"),
                    react_1.default.createElement(NotchedTag_1.default, { size: 'small', tagStyle: 'style-three', type: 'filled', style: style }, "LI"))));
        },
    });
    return react_1.default.createElement(Component, null);
};
exports.Basic = Basic;
