// lucid-ui/dist/cjs/components/ExpanderPanel/ExpanderPanel.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BasicWithOnRestCallback = exports.NoPadding = exports.Basic = void 0;
var lodash_1 = __importDefault(require("lodash"));
var react_1 = __importDefault(require("react"));
var ExpanderPanel_1 = __importDefault(require("./ExpanderPanel"));
exports.default = {
    title: 'Layout/ExpanderPanel',
    component: ExpanderPanel_1.default,
    parameters: {
        docs: {
            description: {
                component: ExpanderPanel_1.default.peek.description,
            },
        },
    },
    args: ExpanderPanel_1.default.defaultProps,
};
/* Basic */
var Basic = function (args) {
    return (react_1.default.createElement(ExpanderPanel_1.default, __assign({}, args),
        react_1.default.createElement(ExpanderPanel_1.default.Header, null, "Show More"),
        lodash_1.default.times(100, function (n) { return (react_1.default.createElement("div", { key: n }, lodash_1.default.repeat('-', 75 * Math.sin(n / 5)))); })));
};
exports.Basic = Basic;
/* No Padding */
var NoPadding = function (args) {
    return (react_1.default.createElement(ExpanderPanel_1.default, __assign({}, args, { hasPadding: false }),
        react_1.default.createElement(ExpanderPanel_1.default.Header, null, "Show More"),
        lodash_1.default.times(100, function (n) { return (react_1.default.createElement("div", { key: n }, lodash_1.default.repeat('-', 75 * Math.sin(n / 5)))); })));
};
exports.NoPadding = NoPadding;
/* Basic With On Rest Callback */
var BasicWithOnRestCallback = function (args) {
    var onRest = function () {
        alert('A big ball of wibbly wobbly, timey wimey stuff');
    };
    return (react_1.default.createElement(ExpanderPanel_1.default, __assign({}, args, { onRest: onRest }),
        react_1.default.createElement(ExpanderPanel_1.default.Header, null, "Show More"),
        lodash_1.default.times(100, function (n) { return (react_1.default.createElement("div", { key: n }, lodash_1.default.repeat('-', 75 * Math.sin(n / 5)))); })));
};
exports.BasicWithOnRestCallback = BasicWithOnRestCallback;
