// lucid-ui/dist/cjs/components/LoadingMessage/LoadingMessage.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomBody = exports.CustomTitle = exports.Basic = void 0;
var react_1 = __importDefault(require("react"));
var LoadingMessage_1 = __importDefault(require("./LoadingMessage"));
exports.default = {
    title: 'Loading Message/LoadingMessage',
    component: LoadingMessage_1.default,
    parameters: {
        docs: {
            description: {
                component: LoadingMessage_1.default.peek.description,
            },
        },
        layout: 'centered',
    },
};
var Basic = function (args) {
    return react_1.default.createElement(LoadingMessage_1.default, null);
};
exports.Basic = Basic;
var CustomTitle = function (args) {
    return react_1.default.createElement(LoadingMessage_1.default, __assign({}, args));
};
exports.CustomTitle = CustomTitle;
exports.CustomTitle.args = {
    Title: 'A custom title',
};
var CustomBody = function (args) {
    return react_1.default.createElement(LoadingMessage_1.default, __assign({}, args));
};
exports.CustomBody = CustomBody;
exports.CustomBody.args = {
    Body: 'A custom body',
};
