// lucid-ui/dist/cjs/components/LoadingMessage/LoadingMessage.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoadingMessage = void 0;
/* eslint-disable react/prop-types */
var lodash_1 = __importStar(require("lodash"));
var react_1 = __importDefault(require("react"));
var prop_types_1 = __importDefault(require("prop-types"));
var style_helpers_1 = require("../../util/style-helpers");
var component_types_1 = require("../../util/component-types");
var LoadingIcon_1 = __importDefault(require("../Icon/LoadingIcon/LoadingIcon"));
var cx = style_helpers_1.lucidClassNames.bind('&-LoadingMessage');
var any = prop_types_1.default.any, node = prop_types_1.default.node, string = prop_types_1.default.string;
var LoadingMessageIcon = function (_props) { return null; };
LoadingMessageIcon.displayName = 'LoadingMessage.Icon';
LoadingMessageIcon.peek = {
    description: "Renders the `Icon` element passed in.",
};
LoadingMessageIcon.propName = 'Icon';
LoadingMessageIcon.propTypes = {
    description: string,
    children: any,
};
var LoadingMessageTitle = function (_props) { return null; };
LoadingMessageTitle.displayName = 'LoadingMessage.Title';
LoadingMessageTitle.peek = {
    description: "Renders an `<h3>` that represents the title of the `LoadingMessage`.  Defaults to the string \"Loading\".",
};
LoadingMessageTitle.propName = 'Title';
LoadingMessageTitle.propTypes = {
    description: string,
    children: any,
};
var LoadingMessageBody = function (_props) { return null; };
LoadingMessageBody.displayName = 'LoadingMessage.Body';
LoadingMessageBody.peek = {
    description: "Renders a `<span>` that represents the body of the `LoadingMessage`.",
};
LoadingMessageBody.propName = 'Body';
LoadingMessageBody.propTypes = {
    description: string,
    children: any,
};
var LoadingMessage = function (props) {
    var className = props.className, passThroughs = __rest(props, ["className"]);
    var Icon = exports.LoadingMessage.Icon, Title = exports.LoadingMessage.Title, Body = exports.LoadingMessage.Body;
    var defaultTitle = 'Loading';
    var iconElement = (0, component_types_1.getFirst)(props, Icon);
    var iconChildren = lodash_1.default.get(iconElement, 'props.children', react_1.default.createElement(LoadingIcon_1.default, null));
    var titleElement = (0, component_types_1.getFirst)(props, Title);
    var titleChildren = lodash_1.default.get(titleElement, 'props.children');
    var bodyElement = (0, component_types_1.getFirst)(props, Body);
    var bodyChildren = lodash_1.default.get(bodyElement, 'props.children', null);
    return (react_1.default.createElement("div", __assign({}, (0, lodash_1.omit)(passThroughs, [
        'className',
        'children',
        'Icon',
        'Title',
        'Body',
        'initialState',
        'callbackId',
    ]), { className: cx('&', { '&-no-content': lodash_1.default.isNull(titleChildren) && !bodyChildren }, className) }),
        iconChildren,
        !lodash_1.default.isNull(titleChildren) && (react_1.default.createElement("h3", { className: cx('&-title') }, titleChildren || defaultTitle)),
        bodyChildren && react_1.default.createElement("span", { className: cx('&-body') }, bodyChildren)));
};
exports.LoadingMessage = LoadingMessage;
exports.LoadingMessage.Icon = LoadingMessageIcon;
exports.LoadingMessage.Title = LoadingMessageTitle;
exports.LoadingMessage.Body = LoadingMessageBody;
exports.LoadingMessage._isPrivate = true;
exports.LoadingMessage.displayName = 'LoadingMessage';
exports.LoadingMessage.peek = {
    description: "A loading message.",
    categories: ['communication'],
    madeFrom: ['LoadingIcon'],
};
exports.LoadingMessage.propTypes = {
    /**
        Class names that are appended to the defaults.
    */
    className: string,
    /**
        Any valid React children.
    */
    children: node,
    /**
        Custom Icon element (alias for \`LoadingMessage.Icon\`)
    */
    Icon: node,
    /**
        Custom Title element (alias for \`LoadingMessage.Title\`)
    */
    Title: node,
    /**
        Custom Body element (alias for \`LoadingMessage.Body\`)
    */
    Body: node,
};
exports.default = exports.LoadingMessage;
