// lucid-ui/dist/cjs/components/RadioButton/RadioButton.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.States = exports.Interactive = void 0;
var react_1 = __importStar(require("react"));
var create_react_class_1 = __importDefault(require("create-react-class"));
var RadioButton_1 = __importDefault(require("./RadioButton"));
exports.default = {
    title: 'Controls/RadioButton',
    component: RadioButton_1.default,
    parameters: {
        docs: {
            description: {
                component: RadioButton_1.default.peek.description,
            },
        },
    },
};
/* Interactive */
var Interactive = function (args) {
    var style = {
        listStyleType: 'none',
        display: 'flex',
        alignItems: 'center',
    };
    var _a = (0, react_1.useState)(false), selected = _a[0], setSelected = _a[1];
    var handleSelected = function (value) {
        setSelected(!value);
    };
    return (react_1.default.createElement("ul", null,
        react_1.default.createElement("li", { style: style },
            react_1.default.createElement("label", null, "Enabled"),
            react_1.default.createElement(RadioButton_1.default, __assign({}, args, { isSelected: selected, name: 'interactive-radio-buttons', onSelect: function () { return handleSelected(selected); }, tabIndex: 20 })))));
};
exports.Interactive = Interactive;
/* States */
var States = function (args) {
    var style = {
        display: 'flex',
        alignItems: 'center',
    };
    var Component = (0, create_react_class_1.default)({
        render: function () {
            return (react_1.default.createElement("ul", null,
                react_1.default.createElement("li", { style: style },
                    react_1.default.createElement("label", null, "Unselected"),
                    react_1.default.createElement(RadioButton_1.default, __assign({}, args, { tabIndex: 20 }))),
                react_1.default.createElement("li", { style: style },
                    react_1.default.createElement("label", null, "Selected"),
                    react_1.default.createElement(RadioButton_1.default, __assign({}, args, { isSelected: true, tabIndex: 21 }))),
                react_1.default.createElement("li", { style: style },
                    react_1.default.createElement("label", null, "Disabled"),
                    react_1.default.createElement(RadioButton_1.default, __assign({}, args, { isDisabled: true, tabIndex: 22 }))),
                react_1.default.createElement("li", { style: style },
                    react_1.default.createElement("label", null, "Disabled & selected"),
                    react_1.default.createElement(RadioButton_1.default, __assign({}, args, { isDisabled: true, isSelected: true, tabIndex: 23 })))));
        },
    });
    return react_1.default.createElement(Component, null);
};
exports.States = States;
