// lucid-ui/dist/cjs/components/Collapsible/Collapsible.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ToggleWithOnRestCallback = exports.RenderChildrenOnCollapse = exports.UnmountChildrenOnCollapse = exports.NotAnimated = exports.ToggleExpand = void 0;
var react_1 = __importStar(require("react"));
var Collapsible_1 = __importDefault(require("./Collapsible"));
var Button_1 = __importDefault(require("../Button/Button"));
var SingleSelect_1 = __importDefault(require("../SingleSelect/SingleSelect"));
exports.default = {
    title: 'Utility/Collapsible',
    component: Collapsible_1.default,
    parameters: {
        docs: {
            description: {
                component: Collapsible_1.default.peek.description,
            },
        },
    },
};
/* Toggle Expand */
var ToggleExpand = function (args) {
    var _a = (0, react_1.useState)(true), isExpanded = _a[0], setIsExpanded = _a[1];
    var handleToggle = function () {
        setIsExpanded(!isExpanded);
    };
    return (react_1.default.createElement("section", null,
        react_1.default.createElement(Button_1.default, { onClick: handleToggle }, "toggle"),
        react_1.default.createElement(Collapsible_1.default, __assign({}, args, { isExpanded: isExpanded, style: { background: '#f4f2f2', padding: 10 } }),
            react_1.default.createElement("p", null, "Kitsch aesthetic gluten-free bitters affogato, you probably haven't heard of them DIY in cornhole sunt meditation man braid cardigan. Selfies stumptown sriracha, small batch williamsburg synth organic aute mlkshk beard venmo normcore XOXO freegan chambray. Freegan delectus next level nostrud intelligentsia swag. Pabst delectus selfies kale chips, duis pariatur locavore iPhone. Chia 3 wolf moon tofu messenger bag pop-up intelligentsia. Pinterest excepteur yuccie four dollar toast DIY. Plaid drinking vinegar locavore id mustache."),
            react_1.default.createElement("p", null, "Flexitarian gentrify taxidermy, bitters chambray swag etsy brunch kogi pork belly authentic. Direct trade put a bird on it ad bespoke, occaecat marfa cliche nisi aliquip kale chips synth readymade. Odio austin keytar pour-over fugiat. Pickled everyday carry laborum sartorial pinterest labore, organic voluptate banh mi kitsch. Leggings austin thundercats cliche. Neutra typewriter sapiente, intelligentsia semiotics +1 esse ex authentic aesthetic banh mi drinking vinegar knausgaard. Id tousled culpa, ut chia cillum actually fashion axe cronut forage assumenda tilde ramps."))));
};
exports.ToggleExpand = ToggleExpand;
/* Not Animated */
var NotAnimated = function (args) {
    var _a = (0, react_1.useState)(true), isExpanded = _a[0], setIsExpanded = _a[1];
    var handleToggle = function () {
        setIsExpanded(!isExpanded);
    };
    return (react_1.default.createElement("section", null,
        react_1.default.createElement(Button_1.default, { onClick: handleToggle }, "toggle"),
        react_1.default.createElement(Collapsible_1.default, __assign({}, args, { isAnimated: false, isExpanded: isExpanded, style: { background: '#f4f2f2', padding: 10 } }),
            react_1.default.createElement("p", null, "Kitsch aesthetic gluten-free bitters affogato, you probably haven't heard of them DIY in cornhole sunt meditation man braid cardigan. Selfies stumptown sriracha, small batch williamsburg synth organic aute mlkshk beard venmo normcore XOXO freegan chambray. Freegan delectus next level nostrud intelligentsia swag. Pabst delectus selfies kale chips, duis pariatur locavore iPhone. Chia 3 wolf moon tofu messenger bag pop-up intelligentsia. Pinterest excepteur yuccie four dollar toast DIY. Plaid drinking vinegar locavore id mustache."),
            react_1.default.createElement("p", null, "Flexitarian gentrify taxidermy, bitters chambray swag etsy brunch kogi pork belly authentic. Direct trade put a bird on it ad bespoke, occaecat marfa cliche nisi aliquip kale chips synth readymade. Odio austin keytar pour-over fugiat. Pickled everyday carry laborum sartorial pinterest labore, organic voluptate banh mi kitsch. Leggings austin thundercats cliche. Neutra typewriter sapiente, intelligentsia semiotics +1 esse ex authentic aesthetic banh mi drinking vinegar knausgaard. Id tousled culpa, ut chia cillum actually fashion axe cronut forage assumenda tilde ramps."))));
};
exports.NotAnimated = NotAnimated;
/* Unmount Children On Collapse */
var UnmountChildrenOnCollapse = function (args) {
    var _a = (0, react_1.useState)(true), isExpanded = _a[0], setIsExpanded = _a[1];
    var handleToggle = function () {
        setIsExpanded(!isExpanded);
    };
    return (react_1.default.createElement("section", null,
        react_1.default.createElement(Button_1.default, { onClick: handleToggle }, "toggle"),
        react_1.default.createElement(Collapsible_1.default, __assign({}, args, { isExpanded: isExpanded, style: { background: '#f4f2f2', padding: 10 } }),
            react_1.default.createElement("p", null, "Kitsch aesthetic gluten-free bitters affogato, you probably haven't heard of them DIY in cornhole sunt meditation man braid cardigan. Selfies stumptown sriracha, small batch williamsburg synth organic aute mlkshk beard venmo normcore XOXO freegan chambray. Freegan delectus next level nostrud intelligentsia swag. Pabst delectus selfies kale chips, duis pariatur locavore iPhone. Chia 3 wolf moon tofu messenger bag pop-up intelligentsia. Pinterest excepteur yuccie four dollar toast DIY. Plaid drinking vinegar locavore id mustache."),
            react_1.default.createElement(SingleSelect_1.default, { DropMenu: { isExpanded: true } },
                react_1.default.createElement(SingleSelect_1.default.Option, null, "Will"),
                react_1.default.createElement(SingleSelect_1.default.Option, null, "Not"),
                react_1.default.createElement(SingleSelect_1.default.Option, null, "Remain")),
            react_1.default.createElement("p", null, "Flexitarian gentrify taxidermy, bitters chambray swag etsy brunch kogi pork belly authentic. Direct trade put a bird on it ad bespoke, occaecat marfa cliche nisi aliquip kale chips synth readymade. Odio austin keytar pour-over fugiat. Pickled everyday carry laborum sartorial pinterest labore, organic voluptate banh mi kitsch. Leggings austin thundercats cliche. Neutra typewriter sapiente, intelligentsia semiotics +1 esse ex authentic aesthetic banh mi drinking vinegar knausgaard. Id tousled culpa, ut chia cillum actually fashion axe cronut forage assumenda tilde ramps."))));
};
exports.UnmountChildrenOnCollapse = UnmountChildrenOnCollapse;
/* Render Children On Collapse */
var RenderChildrenOnCollapse = function (args) {
    var _a = (0, react_1.useState)(true), isExpanded = _a[0], setIsExpanded = _a[1];
    var handleToggle = function () {
        setIsExpanded(!isExpanded);
    };
    return (react_1.default.createElement("section", null,
        react_1.default.createElement(Button_1.default, { onClick: handleToggle }, "toggle"),
        react_1.default.createElement(Collapsible_1.default, __assign({}, args, { isMountControlled: false, isExpanded: isExpanded, style: { background: '#f4f2f2', padding: 10 } }),
            react_1.default.createElement("p", null, "Kitsch aesthetic gluten-free bitters affogato, you probably haven't heard of them DIY in cornhole sunt meditation man braid cardigan. Selfies stumptown sriracha, small batch williamsburg synth organic aute mlkshk beard venmo normcore XOXO freegan chambray. Freegan delectus next level nostrud intelligentsia swag. Pabst delectus selfies kale chips, duis pariatur locavore iPhone. Chia 3 wolf moon tofu messenger bag pop-up intelligentsia. Pinterest excepteur yuccie four dollar toast DIY. Plaid drinking vinegar locavore id mustache."),
            react_1.default.createElement("div", { style: { display: 'flex', justifyContent: 'center' } },
                react_1.default.createElement(SingleSelect_1.default, { DropMenu: { isExpanded: true } },
                    react_1.default.createElement(SingleSelect_1.default.Option, null, "Will"),
                    react_1.default.createElement(SingleSelect_1.default.Option, null, "Stay"),
                    react_1.default.createElement(SingleSelect_1.default.Option, null, "Rendered"))),
            react_1.default.createElement("p", null, "Flexitarian gentrify taxidermy, bitters chambray swag etsy brunch kogi pork belly authentic. Direct trade put a bird on it ad bespoke, occaecat marfa cliche nisi aliquip kale chips synth readymade. Odio austin keytar pour-over fugiat. Pickled everyday carry laborum sartorial pinterest labore, organic voluptate banh mi kitsch. Leggings austin thundercats cliche. Neutra typewriter sapiente, intelligentsia semiotics +1 esse ex authentic aesthetic banh mi drinking vinegar knausgaard. Id tousled culpa, ut chia cillum actually fashion axe cronut forage assumenda tilde ramps."))));
};
exports.RenderChildrenOnCollapse = RenderChildrenOnCollapse;
/* Toggle With On Rest Callback */
var ToggleWithOnRestCallback = function (args) {
    var onRest = function () {
        alert('I am and always will be the optimist. The hoper of far-flung hopes and the dreamer of improbable dreams.');
    };
    var _a = (0, react_1.useState)(false), isExpanded = _a[0], setIsExpanded = _a[1];
    var handleToggle = function () {
        setIsExpanded(!isExpanded);
    };
    return (react_1.default.createElement("section", null,
        react_1.default.createElement(Button_1.default, { onClick: handleToggle }, "toggle"),
        react_1.default.createElement(Collapsible_1.default, __assign({}, args, { isExpanded: isExpanded, style: { background: '#f4f2f2', padding: 10 }, onRest: onRest }),
            react_1.default.createElement("p", null, "Everybody knows that everybody dies. But not every day. Not today. Some days are special. Some days are so, so blessed. Some days, nobody dies at all. Now and then, every once in a very long while, every day in a million days, when the wind stands fair and the Doctor comes to call, everybody lives."))));
};
exports.ToggleWithOnRestCallback = ToggleWithOnRestCallback;
exports.ToggleWithOnRestCallback.storyName = 'Toggle With OnRest Callback';
