// lucid-ui/dist/cjs/components/TextField/TextField.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Disabled = exports.Multiline = exports.Debounced = exports.OnSubmit = exports.Plain = exports.Basic = void 0;
var react_1 = __importStar(require("react"));
var TextField_1 = __importDefault(require("./TextField"));
var Button_1 = __importDefault(require("../Button/Button"));
exports.default = {
    title: 'Controls/TextField',
    component: TextField_1.default,
    parameters: {
        docs: {
            description: {
                component: TextField_1.default.peek.description,
            },
        },
    },
};
var style = {
    marginBottom: '10px',
};
var Basic = function (args) {
    var _a = (0, react_1.useState)(''), value = _a[0], setValue = _a[1];
    return (react_1.default.createElement("div", null,
        react_1.default.createElement(TextField_1.default, __assign({}, args, { style: style, placeholder: 'default', value: value, onChange: function (value) { return setValue(value); } }))));
};
exports.Basic = Basic;
var Plain = function (args) {
    return (react_1.default.createElement("div", null,
        react_1.default.createElement(TextField_1.default, __assign({}, args, { style: style, placeholder: 'Plain Textfield example' })),
        react_1.default.createElement(TextField_1.default, __assign({}, args, { isMultiLine: true, rows: 5, style: style, placeholder: 'Plain Textfield multiline example' }))));
};
exports.Plain = Plain;
var OnSubmit = function (args) {
    var _a = (0, react_1.useState)('Enter some text in and hit enter'), value = _a[0], setValue = _a[1];
    return (react_1.default.createElement("div", null,
        react_1.default.createElement(TextField_1.default, __assign({}, args, { style: style, value: value, onSubmit: function (value) { return setValue(value); } })),
        react_1.default.createElement("div", { style: { marginTop: '10px', marginLeft: '10px' } },
            "state.value: ",
            value)));
};
exports.OnSubmit = OnSubmit;
var Debounced = function (args) {
    var _a = (0, react_1.useState)('foo'), value = _a[0], setValue = _a[1];
    return (react_1.default.createElement("div", null,
        react_1.default.createElement(TextField_1.default, __assign({}, args, { style: style, value: value, onChangeDebounced: function (value) { return setValue(value); } })),
        react_1.default.createElement("div", { style: { marginBottom: '10px', marginLeft: '10px' } },
            "state.value: ",
            value),
        react_1.default.createElement(Button_1.default, __assign({}, Button_1.default.defaultProps, { onClick: function () {
                setValue('foo');
            } }), "Set TextField to \"foo\"")));
};
exports.Debounced = Debounced;
var Multiline = function (args) {
    var _a = (0, react_1.useState)(''), value = _a[0], setValue = _a[1];
    return (react_1.default.createElement("div", null,
        react_1.default.createElement(TextField_1.default, __assign({}, args, { placeholder: 'default', style: style, value: value, onChange: function (value) { return setValue(value); } })),
        react_1.default.createElement(TextField_1.default, __assign({}, args, { isMultiLine: true, placeholder: 'isMultiline', style: style, value: value, onChange: function (value) { return setValue(value); } })),
        react_1.default.createElement("div", { style: { marginBottom: '10px', marginLeft: '10px' } },
            "state.value: ",
            value),
        react_1.default.createElement(Button_1.default, __assign({}, Button_1.default.defaultProps, { onClick: function () {
                setValue('foo');
            } }), "Set TextField to \"foo\"")));
};
exports.Multiline = Multiline;
var Disabled = function (args) {
    var _a = (0, react_1.useState)('foo'), value = _a[0], setValue = _a[1];
    return (react_1.default.createElement("div", null,
        react_1.default.createElement(TextField_1.default, __assign({}, args, { placeholder: 'disabled', isDisabled: true, style: style, value: value, onChange: function (value) { return setValue(value); } }))));
};
exports.Disabled = Disabled;
