// lucid-ui/dist/cjs/components/Paginator/Paginator.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectCounts = exports.Disabled = exports.PageSizeSelector = exports.Basic = void 0;
var addons_1 = require("@storybook/addons");
var react_1 = __importDefault(require("react"));
var Paginator_1 = __importDefault(require("./Paginator"));
exports.default = {
    title: 'Navigation/Paginator',
    component: Paginator_1.default,
    parameters: {
        docs: {
            description: {
                component: Paginator_1.default.description,
            },
        },
    },
};
var Basic = function (args) {
    var _a = (0, addons_1.useState)(0), pageIndex = _a[0], setPageIndex = _a[1];
    var _b = (0, addons_1.useState)(0), pageSizeIndex = _b[0], setPageSizeIndex = _b[1];
    var onPageSelect = function (pageIndex) {
        setPageIndex(pageIndex);
    };
    var onPageSizeSelect = function (pageSizeIndex) {
        setPageSizeIndex(pageSizeIndex);
    };
    return (react_1.default.createElement(react_1.default.Fragment, null,
        react_1.default.createElement("p", null,
            "Page Index: ",
            pageIndex),
        react_1.default.createElement("p", null,
            "Page Size Index: ",
            pageSizeIndex),
        react_1.default.createElement(Paginator_1.default, __assign({}, args, { selectedPageIndex: pageIndex, onPageSelect: onPageSelect, selectedPageSizeIndex: pageSizeIndex, onPageSizeSelect: onPageSizeSelect }))));
};
exports.Basic = Basic;
exports.Basic.args = {
    hasPageSizeSelector: true,
    totalCount: 500,
    isDisabled: false,
};
var PageSizeSelector = function (args) { return (react_1.default.createElement(Paginator_1.default, __assign({}, args))); };
exports.PageSizeSelector = PageSizeSelector;
exports.PageSizeSelector.args = {
    hasPageSizeSelector: true,
    totalCount: 500,
};
var Disabled = function (args) { return (react_1.default.createElement(Paginator_1.default, __assign({}, args))); };
exports.Disabled = Disabled;
exports.Disabled.args = {
    isDisabled: true,
};
var ObjectCounts = function (args) { return (react_1.default.createElement(Paginator_1.default, __assign({}, args))); };
exports.ObjectCounts = ObjectCounts;
exports.ObjectCounts.args = {
    hasPageSizeSelector: true,
    totalCount: 100,
    showTotalObjects: true,
    objectLabel: 'Object',
    objectLabelPlural: 'Objects',
};
