// lucid-ui/dist/cjs/components/Switch/Switch.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SelectedDisabledIncludeExclude = exports.DisabledIncludeExclude = exports.SelectedIncludeExclude = exports.IncludeExclude = exports.DisabledSelected = exports.Disabled = exports.Selected = exports.Basic = void 0;
var react_1 = __importStar(require("react"));
var Switch_1 = __importDefault(require("./Switch"));
exports.default = {
    title: 'Controls/Switch',
    component: Switch_1.default,
    parameters: {
        docs: {
            description: {
                component: Switch_1.default.peek.description,
            },
        },
    },
    args: Switch_1.default.defaultProps,
};
var Basic = function (args) {
    var _a = (0, react_1.useState)(true), selected = _a[0], setSelected = _a[1];
    var handleSelect = function () {
        setSelected(!selected);
    };
    return react_1.default.createElement(Switch_1.default, __assign({}, args, { onSelect: handleSelect, isSelected: selected }));
};
exports.Basic = Basic;
var Selected = function (args) { return (react_1.default.createElement(Switch_1.default, __assign({}, args, { title: 'Selected', isSelected: true }))); };
exports.Selected = Selected;
var Disabled = function (args) { return (react_1.default.createElement(Switch_1.default, __assign({}, args, { title: 'Disabled', isDisabled: true }))); };
exports.Disabled = Disabled;
var DisabledSelected = function (args) { return (react_1.default.createElement(Switch_1.default, __assign({}, args, { title: 'Disabled and Selected', isSelected: true, isDisabled: true }))); };
exports.DisabledSelected = DisabledSelected;
var IncludeExclude = function (args) { return (react_1.default.createElement(Switch_1.default, __assign({}, args, { title: 'Include/Exclude', isIncludeExclude: true }))); };
exports.IncludeExclude = IncludeExclude;
var SelectedIncludeExclude = function (args) { return (react_1.default.createElement(Switch_1.default, __assign({}, args, { title: 'Selected Include/Exclude', isIncludeExclude: true, isSelected: true }))); };
exports.SelectedIncludeExclude = SelectedIncludeExclude;
var DisabledIncludeExclude = function (args) { return (react_1.default.createElement(Switch_1.default, __assign({}, args, { title: 'Disabled Include/Exclude', isIncludeExclude: true, isDisabled: true }))); };
exports.DisabledIncludeExclude = DisabledIncludeExclude;
var SelectedDisabledIncludeExclude = function (args) { return (react_1.default.createElement(Switch_1.default, __assign({}, args, { title: 'Selected and Disabled Include/Exclude', isIncludeExclude: true, isSelected: true, isDisabled: true }))); };
exports.SelectedDisabledIncludeExclude = SelectedDisabledIncludeExclude;
