// lucid-ui/dist/cjs/components/TextFieldValidated/TextFieldValidated.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorTypes = exports.Debounced = exports.Basic = void 0;
var react_1 = __importStar(require("react"));
var TextFieldValidated_1 = __importDefault(require("./TextFieldValidated"));
exports.default = {
    title: 'Controls/TextFieldValidated',
    component: TextFieldValidated_1.default,
    parameters: {
        docs: {
            description: {
                component: TextFieldValidated_1.default.peek.description,
            },
        },
    },
};
var style = {
    marginBottom: '10px',
};
var Basic = function (args) {
    return react_1.default.createElement(TextFieldValidated_1.default, __assign({}, args));
};
exports.Basic = Basic;
var Debounced = function (args) {
    var _a = (0, react_1.useState)(''), value = _a[0], setValue = _a[1];
    return (react_1.default.createElement("div", null,
        react_1.default.createElement(TextFieldValidated_1.default, __assign({}, args, { style: style, value: value, onChangeDebounced: function (value) { return setValue(value); }, Error: value === 'foo' ? null : 'Please enter "foo"' }))));
};
exports.Debounced = Debounced;
var ErrorTypes = function (args) {
    var _a = (0, react_1.useState)(''), value = _a[0], setValue = _a[1];
    return (react_1.default.createElement("div", null,
        react_1.default.createElement(TextFieldValidated_1.default, __assign({}, args, { style: style, value: value, onChangeDebounced: function () { }, Error: 'This is an error' })),
        react_1.default.createElement(TextFieldValidated_1.default, __assign({}, args, { style: style, value: value, onChangeDebounced: function () { }, Error: null, Info: 'This is an info' })),
        react_1.default.createElement(TextFieldValidated_1.default, __assign({}, args, { style: style, value: value, onChangeDebounced: function () { }, Error: null, Success: {
                message: 'This is a Success',
            } })),
        react_1.default.createElement(TextFieldValidated_1.default, __assign({}, args, { style: style, value: value, onChangeDebounced: function () { }, Error: null, Success: {
                message: 'This is a disappearing Success',
                disappearing: true,
            } }))));
};
exports.ErrorTypes = ErrorTypes;
