// lucid-ui/dist/cjs/components/Legend/Legend.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Basic = void 0;
var lodash_1 = __importDefault(require("lodash"));
var react_1 = __importDefault(require("react"));
var chartConstants = __importStar(require("../../constants/charts"));
var Legend_1 = __importDefault(require("./Legend"));
var ToolTip_1 = __importDefault(require("../ToolTip/ToolTip"));
exports.default = {
    title: 'Visualizations/Legend',
    component: Legend_1.default,
    parameters: {
        docs: {
            description: {
                component: Legend_1.default.peek.description,
            },
        },
    },
};
/* Basic */
var Basic = function (args) {
    var Item = Legend_1.default.Item;
    return (react_1.default.createElement("div", null,
        react_1.default.createElement(Legend_1.default, __assign({}, args), lodash_1.default.map(chartConstants.PALETTE_7, function (color, i) { return (react_1.default.createElement(Item, { key: color, hasPoint: true, color: color }, "Partner ".concat(i))); })),
        react_1.default.createElement("br", null),
        react_1.default.createElement(Legend_1.default, __assign({}, args), lodash_1.default.map(chartConstants.PALETTE_7, function (color, i) { return (react_1.default.createElement(Item, { key: color, hasPoint: true, pointKind: i, color: color }, "Partner ".concat(i))); })),
        react_1.default.createElement("br", null),
        react_1.default.createElement(Legend_1.default, __assign({}, args), lodash_1.default.map(chartConstants.PALETTE_7, function (color, i) { return (react_1.default.createElement(Item, { key: color, hasLine: true, color: color }, "Partner ".concat(i))); })),
        react_1.default.createElement("br", null),
        react_1.default.createElement(Legend_1.default, __assign({}, args), lodash_1.default.map(chartConstants.PALETTE_7, function (color, i) { return (react_1.default.createElement(Item, { key: color, hasPoint: true, hasLine: true, pointKind: i, color: color }, "Partner ".concat(i))); })),
        react_1.default.createElement("br", null),
        react_1.default.createElement(Legend_1.default, __assign({}, args, { isReversed: true }), lodash_1.default.map(chartConstants.PALETTE_7, function (color, i) { return (react_1.default.createElement(Item, { key: color, hasPoint: true, hasLine: true, pointKind: i, color: color }, "Partner ".concat(i))); })),
        react_1.default.createElement("br", null),
        react_1.default.createElement(Legend_1.default, __assign({}, args, { orient: 'horizontal' }),
            react_1.default.createElement(Item, { hasLine: true, color: chartConstants.COLOR_GOOD }, "Revenue"),
            react_1.default.createElement(Item, { hasLine: true, color: chartConstants.COLOR_BAD }, "Loss"),
            react_1.default.createElement(Item, { hasPoint: true, color: chartConstants.COLOR_0 }, "Partner 0"),
            react_1.default.createElement(Item, { hasPoint: true, color: chartConstants.COLOR_1 }, "Partner 1")),
        react_1.default.createElement("br", null),
        react_1.default.createElement(Legend_1.default, __assign({}, args, { orient: 'horizontal', isReversed: true }),
            react_1.default.createElement(Item, { hasLine: true, color: chartConstants.COLOR_GOOD }, "Revenue"),
            react_1.default.createElement(Item, { hasLine: true, color: chartConstants.COLOR_BAD }, "Loss"),
            react_1.default.createElement(Item, { hasPoint: true, color: chartConstants.COLOR_0 }, "Partner 0"),
            react_1.default.createElement(Item, { hasPoint: true, color: chartConstants.COLOR_1 }, "Partner 1")),
        react_1.default.createElement("br", null),
        react_1.default.createElement(Legend_1.default, __assign({}, args),
            react_1.default.createElement(Item, { hasLine: true, color: chartConstants.COLOR_GOOD }, "Revenue"),
            react_1.default.createElement(Item, { hasLine: true, color: chartConstants.COLOR_BAD }, "Loss"),
            react_1.default.createElement(Item, { hasPoint: true, color: chartConstants.COLOR_0 }, "Partner 0"),
            react_1.default.createElement(Item, { hasPoint: true, color: chartConstants.COLOR_1 }, "Partner 1")),
        react_1.default.createElement("br", null),
        react_1.default.createElement("br", null),
        react_1.default.createElement("br", null),
        react_1.default.createElement(ToolTip_1.default, { isExpanded: true, direction: 'right', alignment: 'end', isLight: true },
            react_1.default.createElement(ToolTip_1.default.Target, null, "Tooltip example"),
            react_1.default.createElement(ToolTip_1.default.Body, null,
                react_1.default.createElement(Legend_1.default, __assign({}, args, { hasBorders: false }),
                    react_1.default.createElement(Item, { hasLine: true, color: chartConstants.COLOR_GOOD }, "Revenue"),
                    react_1.default.createElement(Item, { hasLine: true, color: chartConstants.COLOR_BAD }, "Loss"),
                    react_1.default.createElement(Item, { hasPoint: true, color: chartConstants.COLOR_0 }, "Partner 0"),
                    react_1.default.createElement(Item, { hasPoint: true, color: chartConstants.COLOR_1 }, "Partner 1"))))));
};
exports.Basic = Basic;
