// lucid-ui/dist/esm/components/LoadingSkeletons/BarChartLoadingSkeleton.stories.js
import React from 'react';
import createClass from 'create-react-class';
import BarChartLoadingSkeleton from './BarChartLoadingSkeleton';
export default {
    title: 'Loading Indicator/BarChartLoadingSkeleton',
    component: BarChartLoadingSkeleton,
    parameters: {
        docs: {
            description: {
                component: BarChartLoadingSkeleton.peek.description,
            },
        },
    },
};
/* Basic */
export const Basic = () => {
    const Component = createClass({
        render() {
            return React.createElement(BarChartLoadingSkeleton, { isLoading: true });
        },
    });
    return React.createElement(Component, null);
};
/* Add Header */
export const AddHeader = () => {
    const Component = createClass({
        render() {
            return (React.createElement(BarChartLoadingSkeleton, { isLoading: true, width: 880, header: 'BarChartLoadingSkeleton added header' }));
        },
    });
    return React.createElement(Component, null);
};
AddHeader.storyName = 'AddHeader';
/* Two Rows Three Columns */
export const TwoRowsThreeColumns = () => {
    const Component = createClass({
        render() {
            return (React.createElement("div", null,
                React.createElement(BarChartLoadingSkeleton, { isLoading: true, width: 210, height: 200, numRows: 2, numColumns: 3, marginRight: 100, marginLeft: 0, marginTop: 0, marginBottom: 50 })));
        },
    });
    return React.createElement(Component, null);
};
TwoRowsThreeColumns.storyName = 'TwoRowsThreeColumns';
