// lucid-ui/dist/esm/components/LoadingSkeletons/CardLoadingSkeleton.stories.js
import React from 'react';
import createClass from 'create-react-class';
import CardLoadingSkeleton from './CardLoadingSkeleton';
export default {
    title: 'Loading Indicator/CardLoadingSkeleton',
    component: CardLoadingSkeleton,
    parameters: {
        docs: {
            description: {
                component: CardLoadingSkeleton.peek.description,
            },
        },
    },
};
/* Basic */
export const Basic = () => {
    const Component = createClass({
        render() {
            return React.createElement(CardLoadingSkeleton, { isLoading: true });
        },
    });
    return React.createElement(Component, null);
};
/* Add Header */
export const AddHeader = () => {
    const Component = createClass({
        render() {
            return React.createElement(CardLoadingSkeleton, { isLoading: true, header: 'Added Header' });
        },
    });
    return React.createElement(Component, null);
};
AddHeader.storyName = 'AddHeader';
/* Two Rows Three Columns */
export const TwoRowsThreeColumns = () => {
    const Component = createClass({
        render() {
            return (React.createElement("div", null,
                React.createElement(CardLoadingSkeleton, { isLoading: true, width: 200, height: 50, numRows: 2, numColumns: 3 })));
        },
    });
    return React.createElement(Component, null);
};
TwoRowsThreeColumns.storyName = 'TwoRowsThreeColumns';
