// lucid-ui/dist/esm/components/LoadingSkeletons/ComplexTableLoadingSkeleton.stories.js
import React from 'react';
import createClass from 'create-react-class';
import ComplexTableLoadingSkeleton from './ComplexTableLoadingSkeleton';
export default {
    title: 'Loading Indicator/ComplexTableLoadingSkeleton',
    component: ComplexTableLoadingSkeleton,
    parameters: {
        docs: {
            description: {
                component: ComplexTableLoadingSkeleton.peek.description,
            },
        },
    },
};
/* Basic */
export const Basic = () => {
    const Component = createClass({
        render() {
            return React.createElement(ComplexTableLoadingSkeleton, { isLoading: true });
        },
    });
    return React.createElement(Component, null);
};
/* Add Header */
export const AddHeader = () => {
    const Component = createClass({
        render() {
            return (React.createElement(ComplexTableLoadingSkeleton, { isLoading: true, header: 'Added Header' }));
        },
    });
    return React.createElement(Component, null);
};
AddHeader.storyName = 'AddHeader';
/* Three Rows One Columns */
export const ThreeRowsOneColumns = () => {
    const Component = createClass({
        render() {
            return (React.createElement("div", null,
                React.createElement(ComplexTableLoadingSkeleton, { isLoading: true, width: 860, numRows: 3 })));
        },
    });
    return React.createElement(Component, null);
};
ThreeRowsOneColumns.storyName = 'ThreeRowsOneColumns';
