// lucid-ui/dist/esm/components/LoadingSkeletons/ComplexTableLoadingSkeleton.js
import React from 'react';
import LoadingSkeleton from './LoadingSkeleton';
import { cxBackgroundGray, cxBackgroundNeutral, ComplexTableSvgPath, } from './LoadingSkeletonsSvgUtil';
export const ComplexTableSkeleton = (props) => {
    const { width = 860, height = 92, className } = props;
    return (React.createElement("div", { "data-test-id": 'loadingSkeleton-ComplexTableSkeleton' },
        React.createElement("svg", { "data-test-id": 'loadingSkeleton-ComplexTableSkeleton-svg', width: width, height: height, xmlns: 'http://www.w3.org/2000/svg' },
            React.createElement("g", { fill: 'none', fillRule: 'evenodd' },
                React.createElement("path", { className: cxBackgroundNeutral('&', className), d: 'M0 0h2642v91H0z' }),
                React.createElement("path", { className: cxBackgroundGray('&', className), d: 'M0 90h2642v1H0z' }),
                React.createElement("path", { className: cxBackgroundGray('&', className), d: ComplexTableSvgPath })))));
};
const ComplexTableLoadingSkeleton = (props) => {
    return (React.createElement(LoadingSkeleton, { "data-test-id": 'loadingSkeleton-ComplexTableLoadingSkeleton', Skeleton: ComplexTableSkeleton, ...props }));
};
ComplexTableLoadingSkeleton.displayName = 'ComplexTableLoadingSkeleton';
ComplexTableLoadingSkeleton.peek = {
    description: `A loading indicator wrapper with optional overlay.`,
    notes: {
        overview: `
			A visual indication that a section or component of the interface is loading. Designed to indicate loading data
		`,
        intendedUse: `
			- Use in places where data takes time to load. ComplexTableLoadingSkeleton lets users know that the information they expect to see will appear shortly.		
		`,
        technicalRecommendations: `
			If a page is displaying a lot of data coming from multiple sources, try as best as possible to load the 
			individual parts of the UI, so as not to disrupt the user and block them from interacting with the entire page until all data is loaded.
		`,
    },
    categories: ['Loading Indicator'],
};
export default ComplexTableLoadingSkeleton;
