// lucid-ui/dist/esm/components/LoadingSkeletons/GroupLoadingSkeleton.stories.js
import React from 'react';
import createClass from 'create-react-class';
import GroupLoadingSkeleton from './GroupLoadingSkeleton';
export default {
    title: 'Loading Indicator/GroupLoadingSkeleton',
    component: GroupLoadingSkeleton,
    parameters: {
        docs: {
            description: {
                component: GroupLoadingSkeleton.peek.description,
            },
        },
    },
};
/* Basic */
export const Basic = () => {
    const Component = createClass({
        render() {
            return React.createElement(GroupLoadingSkeleton, { isLoading: true });
        },
    });
    return React.createElement(Component, null);
};
/* Add Header */
export const AddHeader = () => {
    const Component = createClass({
        render() {
            return React.createElement(GroupLoadingSkeleton, { isLoading: true, header: 'Added Header' });
        },
    });
    return React.createElement(Component, null);
};
AddHeader.storyName = 'AddHeader';
/* Two Rows Three Columns */
export const TwoRowsThreeColumns = () => {
    const Component = createClass({
        render() {
            return (React.createElement("div", null,
                React.createElement(GroupLoadingSkeleton, { isLoading: true, width: 250, numRows: 2, numColumns: 3, marginBottom: 30, marginRight: 30 })));
        },
    });
    return React.createElement(Component, null);
};
TwoRowsThreeColumns.storyName = 'TwoRowsThreeColumns';
