// lucid-ui/dist/esm/components/LoadingSkeletons/LineChartLoadingSkeleton.stories.js
import React from 'react';
import createClass from 'create-react-class';
import LineChartLoadingSkeleton from './LineChartLoadingSkeleton';
export default {
    title: 'Loading Indicator/LineChartLoadingSkeleton',
    component: LineChartLoadingSkeleton,
    parameters: {
        docs: {
            description: {
                component: LineChartLoadingSkeleton.peek.description,
            },
        },
    },
};
/* Basic */
export const Basic = () => {
    const Component = createClass({
        render() {
            return React.createElement(LineChartLoadingSkeleton, { isLoading: true });
        },
    });
    return React.createElement(Component, null);
};
/* Add Header */
export const AddHeader = () => {
    const Component = createClass({
        render() {
            return (React.createElement(LineChartLoadingSkeleton, { isLoading: true, header: 'Added Header' }));
        },
    });
    return React.createElement(Component, null);
};
AddHeader.storyName = 'AddHeader';
/* Two Rows Two Columns */
export const TwoRowsTwoColumns = () => {
    const Component = createClass({
        render() {
            return (React.createElement("div", null,
                React.createElement(LineChartLoadingSkeleton, { isLoading: true, width: 400, numRows: 2, numColumns: 2, marginRight: 50, marginBottom: 50 })));
        },
    });
    return React.createElement(Component, null);
};
TwoRowsTwoColumns.storyName = 'TwoRowsTwoColumns';
