// lucid-ui/dist/esm/components/LoadingSkeletons/SimpleTableLoadingSkeleton.stories.js
import React from 'react';
import createClass from 'create-react-class';
import SimpleTableLoadingSkeleton from './SimpleTableLoadingSkeleton';
export default {
    title: 'Loading Indicator/SimpleTableLoadingSkeleton',
    component: SimpleTableLoadingSkeleton,
    parameters: {
        docs: {
            description: {
                component: SimpleTableLoadingSkeleton.peek.description,
            },
        },
    },
};
/* Basic */
export const Basic = () => {
    const Component = createClass({
        render() {
            return React.createElement(SimpleTableLoadingSkeleton, { isLoading: true });
        },
    });
    return React.createElement(Component, null);
};
/* Add Header */
export const AddHeader = () => {
    const Component = createClass({
        render() {
            return (React.createElement(SimpleTableLoadingSkeleton, { isLoading: true, height: 50, header: 'Added Header' }));
        },
    });
    return React.createElement(Component, null);
};
AddHeader.storyName = 'AddHeader';
/* Three Rows One Column */
export const ThreeRowsOneColumn = () => {
    const Component = createClass({
        render() {
            return (React.createElement("div", null,
                React.createElement(SimpleTableLoadingSkeleton, { isLoading: true, height: 70, numRows: 3 })));
        },
    });
    return React.createElement(Component, null);
};
ThreeRowsOneColumn.storyName = 'ThreeRowsOneColumn';
