// lucid-ui/dist/esm/components/LoadingSkeletons/SingleLineLoadingSkeleton.stories.js
import React from 'react';
import createClass from 'create-react-class';
import SingleLineLoadingSkeleton from './SingleLineLoadingSkeleton';
export default {
    title: 'Loading Indicator/SingleLineLoadingSkeleton',
    component: SingleLineLoadingSkeleton,
    parameters: {
        docs: {
            description: {
                component: SingleLineLoadingSkeleton.peek.description,
            },
        },
    },
};
/* Basic */
export const Basic = () => {
    const Component = createClass({
        render() {
            return React.createElement(SingleLineLoadingSkeleton, { isLoading: true });
        },
    });
    return React.createElement(Component, null);
};
/* Add Header */
export const AddHeader = () => {
    const Component = createClass({
        render() {
            return (React.createElement(SingleLineLoadingSkeleton, { isLoading: true, width: 700, height: 50, header: 'Added Header' }));
        },
    });
    return React.createElement(Component, null);
};
AddHeader.storyName = 'AddHeader';
/* Two Rows Three Columns */
export const TwoRowsThreeColumns = () => {
    const Component = createClass({
        render() {
            return (React.createElement("div", null,
                React.createElement(SingleLineLoadingSkeleton, { isLoading: true, width: 250, height: 20, numRows: 2, numColumns: 3, marginRight: 20, marginBottom: 10 })));
        },
    });
    return React.createElement(Component, null);
};
TwoRowsThreeColumns.storyName = 'TwoRowsThreeColumns';
