// lucid-ui/dist/esm/components/LoadingSkeletons/SmallDataTableLoadingSkeleton.stories.js
import React from 'react';
import createClass from 'create-react-class';
import SmallDataTableLoadingSkeleton from './SmallDataTableLoadingSkeleton';
export default {
    title: 'Loading Indicator/SmallDataTableLoadingSkeleton',
    component: SmallDataTableLoadingSkeleton,
    parameters: {
        docs: {
            description: {
                component: SmallDataTableLoadingSkeleton.peek.description,
            },
        },
    },
};
/* Basic */
export const Basic = () => {
    const Component = createClass({
        render() {
            return React.createElement(SmallDataTableLoadingSkeleton, { isLoading: true });
        },
    });
    return React.createElement(Component, null);
};
/* Add Header */
export const AddHeader = () => {
    const Component = createClass({
        render() {
            return (React.createElement(SmallDataTableLoadingSkeleton, { isLoading: true, width: 700, header: 'Added Header' }));
        },
    });
    return React.createElement(Component, null);
};
AddHeader.storyName = 'AddHeader';
/* Two Rows Two Columns */
export const TwoRowsTwoColumns = () => {
    const Component = createClass({
        render() {
            return (React.createElement("div", null,
                React.createElement(SmallDataTableLoadingSkeleton, { isLoading: true, width: 300, numRows: 2, numColumns: 2, marginRight: 100, marginBottom: 30 })));
        },
    });
    return React.createElement(Component, null);
};
TwoRowsTwoColumns.storyName = 'TwoRowsTwoColumns';
