// lucid-ui/dist/esm/components/LoadingSkeletons/TableLoadingSkeleton.stories.js
import React from 'react';
import createClass from 'create-react-class';
import TableLoadingSkeleton from './TableLoadingSkeleton';
export default {
    title: 'Loading Indicator/TableLoadingSkeleton',
    component: TableLoadingSkeleton,
    parameters: {
        docs: {
            description: {
                component: TableLoadingSkeleton.peek.description,
            },
        },
    },
};
/* Basic */
export const Basic = () => {
    const Component = createClass({
        render() {
            return React.createElement(TableLoadingSkeleton, { isLoading: true });
        },
    });
    return React.createElement(Component, null);
};
/* Add Header */
export const AddHeader = () => {
    const Component = createClass({
        render() {
            return React.createElement(TableLoadingSkeleton, { isLoading: true, header: 'Added Header' });
        },
    });
    return React.createElement(Component, null);
};
AddHeader.storyName = 'AddHeader';
/* Two Rows Three Columns */
export const TwoRowsThreeColumns = () => {
    const Component = createClass({
        render() {
            return (React.createElement("div", null,
                React.createElement(TableLoadingSkeleton, { isLoading: true, width: 200, height: 100, numRows: 2, numColumns: 3, header: 'Rows And Columns', marginRight: 50, marginLeft: 0, marginTop: 0, marginBottom: 20 })));
        },
    });
    return React.createElement(Component, null);
};
TwoRowsThreeColumns.storyName = 'TwoRowsThreeColumns';
