// lucid-ui/dist/cjs/components/SplitButton/SplitButton.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Sizes = exports.Up = exports.Disabled = exports.Basic = void 0;
var react_1 = __importDefault(require("react"));
var SplitButton_1 = __importDefault(require("./SplitButton"));
exports.default = {
    title: 'Controls/SplitButton',
    component: SplitButton_1.default,
    parameters: {
        docs: {
            description: {
                component: SplitButton_1.default.peek.description,
            },
        },
    },
    args: SplitButton_1.default.defaultProps,
};
/* Basic */
var Basic = function (args) {
    var style = { marginRight: '20px', height: 100 };
    return (react_1.default.createElement("section", null,
        react_1.default.createElement(SplitButton_1.default, __assign({}, args, { style: style }),
            react_1.default.createElement(SplitButton_1.default.Button, null, "Basic"),
            react_1.default.createElement(SplitButton_1.default.Button, null, "One"),
            react_1.default.createElement(SplitButton_1.default.Button, null, "Two")),
        react_1.default.createElement(SplitButton_1.default, __assign({}, args, { kind: 'primary', style: style }),
            react_1.default.createElement(SplitButton_1.default.Button, null, "Primary"),
            react_1.default.createElement(SplitButton_1.default.Button, null, "One"),
            react_1.default.createElement(SplitButton_1.default.Button, null, "Two"))));
};
exports.Basic = Basic;
/* Disabled */
var Disabled = function (args) {
    return (react_1.default.createElement("section", { style: { height: 100 } },
        react_1.default.createElement(SplitButton_1.default, __assign({}, args),
            react_1.default.createElement(SplitButton_1.default.Button, { isDisabled: true }, "Save"),
            react_1.default.createElement(SplitButton_1.default.Button, { isDisabled: true }, "This action should be disabled"),
            react_1.default.createElement(SplitButton_1.default.Button, null, "This one should be enabled"),
            react_1.default.createElement(SplitButton_1.default.Button, { isDisabled: true }, "This should be disabled, too"))));
};
exports.Disabled = Disabled;
/* Up */
var Up = function (args) {
    return (react_1.default.createElement("section", { style: { height: 100 } },
        react_1.default.createElement(SplitButton_1.default, __assign({}, args, { direction: 'up', kind: 'primary' }),
            react_1.default.createElement(SplitButton_1.default.Button, null, "Save"),
            react_1.default.createElement(SplitButton_1.default.Button, null, "Action 01"),
            react_1.default.createElement(SplitButton_1.default.Button, null, "Here's Another Action"),
            react_1.default.createElement(SplitButton_1.default.Button, null, "And Another Action"))));
};
exports.Up = Up;
/* Sizes */
var Sizes = function (args) {
    var style = { marginRight: '20px', height: 100 };
    return (react_1.default.createElement("section", null,
        react_1.default.createElement(SplitButton_1.default, __assign({}, args, { size: 'large', style: style }),
            react_1.default.createElement(SplitButton_1.default.Button, null, "Large"),
            react_1.default.createElement(SplitButton_1.default.Button, null, "One"),
            react_1.default.createElement(SplitButton_1.default.Button, null, "Two")),
        react_1.default.createElement(SplitButton_1.default, __assign({}, args, { size: 'small', style: style }),
            react_1.default.createElement(SplitButton_1.default.Button, null, "Small"),
            react_1.default.createElement(SplitButton_1.default.Button, null, "One"),
            react_1.default.createElement(SplitButton_1.default.Button, null, "Two")),
        react_1.default.createElement(SplitButton_1.default, __assign({}, args, { size: 'short', style: style }),
            react_1.default.createElement(SplitButton_1.default.Button, null, "Short"),
            react_1.default.createElement(SplitButton_1.default.Button, null, "One"),
            react_1.default.createElement(SplitButton_1.default.Button, null, "Two"))));
};
exports.Sizes = Sizes;
