// lucid-ui/dist/cjs/components/AxisLabel/AxisLabel.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Basic = void 0;
var react_1 = __importDefault(require("react"));
var chartConstants = __importStar(require("../../constants/charts"));
var AxisLabel_1 = __importDefault(require("./AxisLabel"));
exports.default = {
    title: 'Visualizations/AxisLabel',
    component: AxisLabel_1.default,
    parameters: {
        docs: {
            description: {
                component: AxisLabel_1.default.peek.description,
            },
        },
    },
};
/* Basic */
var Basic = function (args) {
    var width = 800;
    var height = 400;
    var margin = { top: 50, right: 50, bottom: 50, left: 50 };
    var innerWidth = width - margin.right - margin.left;
    var innerHeight = height - margin.top - margin.bottom;
    return (react_1.default.createElement("svg", { width: width, height: height },
        react_1.default.createElement("rect", { x: 1, y: 1, width: width - 2, height: height - 2, style: {
                strokeWidth: 1,
                strokeDasharray: '2,2',
                fill: 'none',
                stroke: 'black',
            } }),
        react_1.default.createElement("g", { transform: "translate(".concat(margin.left, ", ").concat(margin.top, ")") },
            react_1.default.createElement("rect", { style: {
                    fill: 'lightgrey',
                }, width: innerWidth, height: innerHeight })),
        react_1.default.createElement("g", { transform: 'translate(0,0)' },
            react_1.default.createElement(AxisLabel_1.default, __assign({}, args, { orient: 'left', width: margin.left, height: height, label: 'Left' }))),
        react_1.default.createElement("g", { transform: "translate(".concat(margin.left + innerWidth, ", 0)") },
            react_1.default.createElement(AxisLabel_1.default, __assign({}, args, { orient: 'right', color: chartConstants.COLOR_0, width: margin.right, height: height, label: 'Right' }))),
        react_1.default.createElement("g", { transform: "translate(0, ".concat(margin.top + innerHeight, ")") },
            react_1.default.createElement(AxisLabel_1.default, __assign({}, args, { orient: 'bottom', color: chartConstants.COLOR_1, width: width, height: margin.bottom, label: 'Bottom' }))),
        react_1.default.createElement("g", { transform: 'translate(0, 0)' },
            react_1.default.createElement(AxisLabel_1.default, __assign({}, args, { orient: 'top', color: chartConstants.COLOR_2, width: width, height: margin.top, label: 'Top' })))));
};
exports.Basic = Basic;
