// lucid-ui/dist/esm/components/Paginator/Paginator.stories.js
import { useState } from '@storybook/addons';
import React from 'react';
import Paginator from './Paginator';
export default {
    title: 'Navigation/Paginator',
    component: Paginator,
    parameters: {
        docs: {
            description: {
                component: Paginator.description,
            },
        },
    },
};
export const Basic = (args) => {
    const [pageIndex, setPageIndex] = useState(0);
    const [pageSizeIndex, setPageSizeIndex] = useState(0);
    const onPageSelect = (pageIndex) => {
        setPageIndex(pageIndex);
    };
    const onPageSizeSelect = (pageSizeIndex) => {
        setPageSizeIndex(pageSizeIndex);
    };
    return (React.createElement(React.Fragment, null,
        React.createElement("p", null,
            "Page Index: ",
            pageIndex),
        React.createElement("p", null,
            "Page Size Index: ",
            pageSizeIndex),
        React.createElement(Paginator, { ...args, selectedPageIndex: pageIndex, onPageSelect: onPageSelect, selectedPageSizeIndex: pageSizeIndex, onPageSizeSelect: onPageSizeSelect })));
};
Basic.args = {
    hasPageSizeSelector: true,
    totalCount: 500,
    isDisabled: false,
};
export const PageSizeSelector = (args) => (React.createElement(Paginator, { ...args }));
PageSizeSelector.args = {
    hasPageSizeSelector: true,
    totalCount: 500,
};
export const Disabled = (args) => (React.createElement(Paginator, { ...args }));
Disabled.args = {
    isDisabled: true,
};
export const ObjectCounts = (args) => (React.createElement(Paginator, { ...args }));
ObjectCounts.args = {
    hasPageSizeSelector: true,
    totalCount: 100,
    showTotalObjects: true,
    objectLabel: 'Object',
    objectLabelPlural: 'Objects',
};
