// lucid-ui/dist/cjs/components/Underline/Underline.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WithStringMatch = exports.WithRegexMatch = exports.WithDefaults = void 0;
var react_1 = __importDefault(require("react"));
var Underline_1 = __importDefault(require("./Underline"));
exports.default = {
    title: 'Controls/Underline',
    component: Underline_1.default,
    parameters: {
        docs: {
            description: {
                component: Underline_1.default.peek.description,
            },
        },
    },
};
/* With Defaults */
var WithDefaults = function (args) {
    return react_1.default.createElement(Underline_1.default, __assign({}, args), "foo bar baz");
};
exports.WithDefaults = WithDefaults;
/* With Regex Match */
var WithRegexMatch = function (args) {
    return (react_1.default.createElement(Underline_1.default, __assign({}, args, { match: /foo?/i }), "foo bar baz"));
};
exports.WithRegexMatch = WithRegexMatch;
/* With String Match */
var WithStringMatch = function (args) {
    return (react_1.default.createElement(Underline_1.default, __assign({}, args, { match: 'bar' }), "foo bar baz"));
};
exports.WithStringMatch = WithStringMatch;
