// lucid-ui/dist/cjs/components/ScrollTable/ScrollTable.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScrollTable = void 0;
var lodash_1 = require("lodash");
var react_1 = __importDefault(require("react"));
var prop_types_1 = __importDefault(require("prop-types"));
var style_helpers_1 = require("../../util/style-helpers");
var Table_1 = __importDefault(require("../Table/Table"));
var cx = style_helpers_1.lucidClassNames.bind('&-ScrollTable');
var object = prop_types_1.default.object, string = prop_types_1.default.string, bool = prop_types_1.default.bool, node = prop_types_1.default.node, number = prop_types_1.default.number, oneOfType = prop_types_1.default.oneOfType;
var defaultProps = {
    hasWordWrap: false,
    hasBorder: false,
};
var ScrollTable = function (props) {
    var children = props.children, className = props.className, style = props.style, tableWidth = props.tableWidth, hasWordWrap = props.hasWordWrap, hasBorder = props.hasBorder, passThroughs = __rest(props, ["children", "className", "style", "tableWidth", "hasWordWrap", "hasBorder"]);
    return (react_1.default.createElement("div", { className: cx('&', {
            '&-has-border': hasBorder,
        }, className), style: style },
        react_1.default.createElement(Table_1.default, __assign({}, (0, lodash_1.omit)(passThroughs, [
            'children',
            'className',
            'style',
            'tableWidth',
            'hasWordWrap',
            'hasBorder',
            'initialState',
        ]), { style: {
                width: tableWidth,
            }, hasWordWrap: hasWordWrap }), children)));
};
exports.ScrollTable = ScrollTable;
exports.ScrollTable.defaultProps = defaultProps;
exports.ScrollTable.displayName = 'ScrollTable';
exports.ScrollTable.peek = {
    description: "`Table` in a scrollable container.",
    categories: ['table'],
    madeFrom: ['Table'],
};
exports.ScrollTable.Thead = Table_1.default.Thead;
exports.ScrollTable.Tbody = Table_1.default.Tbody;
exports.ScrollTable.Tr = Table_1.default.Tr;
exports.ScrollTable.Th = Table_1.default.Th;
exports.ScrollTable.Td = Table_1.default.Td;
exports.ScrollTable.propTypes = {
    /**
        {Thead, Tbody, Tr, Th, Td} are the child components of Scrolltable, same
        as Table.
    */
    children: node,
    /**
        Class names that are appended to the defaults.
    */
    className: string,
    /**
        Styles that are passed through to the root container.
    */
    style: object,
    /**
        Set the width of the Table inside the scrollable container.
    */
    tableWidth: oneOfType([number, string]),
    /**
        Set the Table contents to not allow word wrapping.
    */
    hasWordWrap: bool,
    /**
        render the table with borders on the outer edge.
    */
    hasBorder: bool,
};
exports.default = exports.ScrollTable;
