// lucid-ui/dist/cjs/components/OverlayWrapper/OverlayWrapper.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Basic = void 0;
var react_1 = __importDefault(require("react"));
var OverlayWrapper_1 = __importDefault(require("./OverlayWrapper"));
var BarChart_1 = __importDefault(require("../BarChart/BarChart"));
exports.default = {
    title: 'Utility/OverlayWrapper',
    component: OverlayWrapper_1.default,
    parameters: {
        docs: {
            description: {
                component: OverlayWrapper_1.default.peek.description,
            },
        },
    },
};
/* Basic */
var Basic = function (args) {
    return (react_1.default.createElement(OverlayWrapper_1.default, __assign({}, args, { isVisible: true }),
        react_1.default.createElement(OverlayWrapper_1.default.Message, null, "Message Goes Here"),
        react_1.default.createElement(BarChart_1.default, { data: [
                { x: '2015-01-01', y: 1 },
                { x: '2015-01-02', y: 2 },
                { x: '2015-01-03', y: 3 },
                { x: '2015-01-04', y: 5 },
            ] })));
};
exports.Basic = Basic;
