// lucid-ui/dist/cjs/components/Accordion/Accordion.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccordionDumb = void 0;
var lodash_1 = __importDefault(require("lodash"));
var react_1 = __importDefault(require("react"));
var prop_types_1 = __importDefault(require("prop-types"));
var style_helpers_1 = require("../../util/style-helpers");
var state_management_1 = require("../../util/state-management");
var ExpanderPanel_1 = require("../ExpanderPanel/ExpanderPanel");
var component_types_1 = require("../../util/component-types");
var reducers = __importStar(require("../Accordion/Accordion.reducers"));
var cx = style_helpers_1.lucidClassNames.bind('&-Accordion');
var func = prop_types_1.default.func, object = prop_types_1.default.object, number = prop_types_1.default.number, string = prop_types_1.default.string;
/** TODO: Remove the nonPassThroughs when the component is converted to a functional component */
var nonPassThroughs = ['selectedIndex', 'onSelect'];
var defaultProps = {
    onSelect: lodash_1.default.noop,
};
var Accordion = function (props) {
    var style = props.style, className = props.className, onSelect = props.onSelect, selectedIndex = props.selectedIndex, passThroughs = __rest(props, ["style", "className", "onSelect", "selectedIndex"]);
    var itemChildProps = lodash_1.default.map((0, component_types_1.findTypes)(props, Accordion.Item), 'props');
    var handleToggle = function (isExpanded, index, event) {
        var selectedIndex = isExpanded ? index : null;
        onSelect(selectedIndex, {
            event: event,
            props: props,
        });
    };
    return (react_1.default.createElement("div", __assign({}, lodash_1.default.omit(passThroughs, nonPassThroughs), { className: cx('&', className), style: style }), lodash_1.default.map(itemChildProps, function (itemChildProp, index) {
        return (react_1.default.createElement(ExpanderPanel_1.ExpanderPanelDumb, __assign({ key: index }, itemChildProp, { className: cx('&-Item', itemChildProp.className), onToggle: function (isExpanded, _a) {
                var event = _a.event;
                return handleToggle(isExpanded, index, event);
            }, isExpanded: !itemChildProp.isDisabled && selectedIndex === index })));
    })));
};
exports.AccordionDumb = Accordion;
Accordion.displayName = 'Accordion';
Accordion.propTypes = {
    /**
        Appended to the component-specific class names set on the root element.
    */
    className: string,
    /**
        Indicates which item is expanded
    */
    selectedIndex: number,
    /**
        Called when the user clicks on the component's header of an item.
    */
    onSelect: func,
    /**
        Passed through to the root element.
    **/
    style: object,
};
Accordion.peek = {
    description: "`Accordion` is a container that renders panels and controls their expansion or collapse.",
    categories: ['layout'],
    madeFrom: ['ExpanderPanel'],
};
Accordion.defaultProps = defaultProps;
Accordion.reducers = reducers;
Accordion.Item = ExpanderPanel_1.ExpanderPanelDumb;
Accordion.Header = ExpanderPanel_1.ExpanderPanelDumb.Header;
exports.default = (0, state_management_1.buildModernHybridComponent)(Accordion, { reducers: reducers });
