// lucid-ui/dist/cjs/components/Bars/Bars.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LogScale = exports.CustomColors = exports.Basic = void 0;
var lodash_1 = __importDefault(require("lodash"));
var react_1 = __importDefault(require("react"));
var Bars_1 = __importDefault(require("../Bars/Bars"));
var chartConstants = __importStar(require("../../constants/charts"));
var d3Scale = __importStar(require("d3-scale"));
exports.default = {
    title: 'Private/Bars',
    component: Bars_1.default,
    parameters: {
        docs: {
            description: {
                component: Bars_1.default.peek.description,
            },
        },
    },
};
/* Basic */
var Basic = function (args) {
    /* eslint-disable comma-spacing */
    var width = 750;
    var height = 400;
    var data = [
        { x: 'one', y0: 1, y1: 2, y2: 3, y3: 5 },
        { x: 'two', y0: 2, y1: 3, y2: 4, y3: 6 },
        { x: 'three', y0: 2, y1: 4, y2: 5, y3: 6 },
        { x: 'four', y0: 3, y1: 6, y2: 7, y3: 7 },
        { x: 'five', y0: 4, y1: 8, y2: 9, y3: 8 },
        { x: 'six', y0: 11, y1: 8, y2: 9, y3: 5 },
    ];
    var yFields = ['y0', 'y1', 'y2', 'y3'];
    var yMax = lodash_1.default.max(lodash_1.default.reduce(yFields, function (acc, field) {
        return acc.concat(lodash_1.default.map(data, field));
    }, []));
    var xScale = d3Scale
        .scaleBand()
        .domain(lodash_1.default.map(data, 'x'))
        .range([0, width])
        .round(true)
        .paddingInner(0.1);
    var yScale = d3Scale
        .scaleLinear()
        .domain([0, yMax])
        .range([height, 0]);
    var style = {
        paddingTop: '9rem',
    };
    return (react_1.default.createElement("div", { style: style },
        react_1.default.createElement("svg", { width: width, height: height },
            react_1.default.createElement(Bars_1.default, __assign({}, args, { data: data, xScale: xScale, yScale: yScale, yFields: yFields, hasToolTips: true }))),
        react_1.default.createElement("svg", { width: width, height: height },
            react_1.default.createElement(Bars_1.default, __assign({}, args, { data: data, xScale: xScale, yScale: yScale, yFields: yFields, isStacked: true, hasToolTips: true })))));
};
exports.Basic = Basic;
/* Custom Colors */
var CustomColors = function (args) {
    /* eslint-disable comma-spacing */
    var width = 750;
    var height = 400;
    var data = [
        { x: 'one', y0: 1, y1: 2, y2: 3, y3: 5 },
        { x: 'two', y0: 2, y1: 3, y2: 4, y3: 6 },
        { x: 'three', y0: 2, y1: 4, y2: 5, y3: 6 },
        { x: 'four', y0: 3, y1: 6, y2: 7, y3: 7 },
        { x: 'five', y0: 4, y1: 8, y2: 9, y3: 8 },
        { x: 'six', y0: 11, y1: 8, y2: 9, y3: 5 },
    ];
    var yFields = ['y0', 'y1', 'y2', 'y3'];
    var yMax = lodash_1.default.max(lodash_1.default.reduce(yFields, function (acc, field) {
        return acc.concat(lodash_1.default.map(data, field));
    }, []));
    var xScale = d3Scale
        .scaleBand()
        .domain(lodash_1.default.map(data, 'x'))
        .range([0, width])
        .round(true)
        .paddingInner(0.1);
    var yScale = d3Scale
        .scaleLinear()
        .domain([0, yMax])
        .range([height, 0]);
    var style = {
        paddingTop: '9rem',
    };
    return (react_1.default.createElement("div", { style: style },
        react_1.default.createElement("svg", { width: width, height: height },
            react_1.default.createElement(Bars_1.default, __assign({}, args, { data: data, xScale: xScale, yScale: yScale, yFields: yFields, colorMap: {
                    y0: chartConstants.COLOR_GOOD,
                    y2: chartConstants.COLOR_BAD,
                }, hasToolTips: true })))));
};
exports.CustomColors = CustomColors;
/* Log Scale */
var LogScale = function (args) {
    var data = [
        { x: 'one', y: 10 },
        { x: 'two', y: 100 },
        { x: 'three', y: 1000 },
        { x: 'four', y: 10000 },
        { x: 'five', y: 100000 },
    ];
    var width = 750;
    var height = 400;
    var yScale = d3Scale.scaleLog().domain([1, 1000000]).range([height, 0]);
    var xScale = d3Scale
        .scaleBand()
        .domain(lodash_1.default.map(data, 'x'))
        .range([0, width]);
    return (react_1.default.createElement("svg", { height: 600, width: 1000 },
        react_1.default.createElement("g", null,
            react_1.default.createElement(Bars_1.default, __assign({}, args, { data: data, xScale: xScale, yScale: yScale }))),
        react_1.default.createElement("g", null, lodash_1.default.map(data, function (item, idx) {
            return (react_1.default.createElement("text", { key: idx, textAnchor: 'middle', x: xScale(item.x) + xScale.bandwidth() / 2, y: yScale(item.y) - 10, fill: 'gray' }, item.y));
        }))));
};
exports.LogScale = LogScale;
