// lucid-ui/dist/cjs/components/Overlay/Overlay.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NoModal = exports.Basic = void 0;
var lodash_1 = __importDefault(require("lodash"));
var react_1 = __importStar(require("react"));
var Overlay_1 = __importDefault(require("./Overlay"));
var Button_1 = __importDefault(require("../Button/Button"));
exports.default = {
    title: 'Utility/Overlay',
    component: Overlay_1.default,
    parameters: {
        docs: {
            description: {
                component: Overlay_1.default.peek.description,
            },
        },
    },
};
/* BASIC */
var Basic = function (args) {
    var _a = (0, react_1.useState)(false), isShown = _a[0], setIsShown = _a[1];
    var handleOpenClose = function (isShown) {
        setIsShown(isShown);
    };
    return (react_1.default.createElement("div", null,
        react_1.default.createElement(Button_1.default, { onClick: lodash_1.default.partial(handleOpenClose, !isShown) }, "Toggle"),
        react_1.default.createElement(Overlay_1.default, __assign({}, args, { isShown: isShown, onEscape: lodash_1.default.partial(handleOpenClose, false), onBackgroundClick: lodash_1.default.partial(handleOpenClose, false) }),
            react_1.default.createElement("div", { style: { color: '#fff' } }, "User cannot interact with the background (except scrolling)."))));
};
exports.Basic = Basic;
/* No Modal */
var NoModal = function (args) {
    var _a = (0, react_1.useState)(false), isShown = _a[0], setIsShown = _a[1];
    var handleOpenClose = function (isShown) {
        setIsShown(isShown);
    };
    return (react_1.default.createElement("div", null,
        react_1.default.createElement(Button_1.default, { onClick: lodash_1.default.partial(handleOpenClose, !isShown) }, "Toggle"),
        react_1.default.createElement(Overlay_1.default, __assign({}, args, { isShown: isShown, isModal: false, onEscape: lodash_1.default.partial(handleOpenClose, false) }),
            react_1.default.createElement("div", { style: {
                    backgroundColor: '#eee',
                    padding: '100px',
                } },
                "User can still interact with the background.",
                react_1.default.createElement(Button_1.default, { onClick: lodash_1.default.partial(handleOpenClose, false), style: { marginLeft: '10px' } }, "Close")))));
};
exports.NoModal = NoModal;
