// lucid-ui/dist/esm/components/Breadcrumb/Breadcrumb.stories.js
import React from 'react';
import Breadcrumb from './Breadcrumb';
import NotchedTag from '../NotchedTag/NotchedTag';
export default {
    title: 'Navigation/Breadcrumb',
    component: Breadcrumb,
    parameters: {
        docs: {
            description: {
                component: Breadcrumb.peek.description,
            },
        },
    },
};
/* Basic */
export const Basic = (args) => {
    const { Item } = Breadcrumb;
    return (React.createElement(Breadcrumb, { ...args },
        React.createElement(Item, { key: 'home' },
            React.createElement("a", { href: '#' }, "Start")),
        React.createElement(Item, { key: 'list' },
            React.createElement("a", { href: '#' }, "List")),
        React.createElement(Item, { key: 'item' },
            React.createElement("a", { href: '#' }, "Item"))));
};
/* Rich */
export const Rich = (args) => {
    const { Item } = Breadcrumb;
    return (React.createElement(Breadcrumb, { ...args },
        React.createElement(Item, { key: 'home' },
            React.createElement("a", { href: '#' },
                React.createElement(NotchedTag, { type: 'stroke', size: 'small', tagStyle: 'style-two' }, "IO"),
                React.createElement("span", null,
                    React.createElement("span", null, "Foo"),
                    React.createElement("span", { style: { fontWeight: 400 } }, "\u00A0(1234)")))),
        React.createElement(Item, { key: 'list' },
            React.createElement("a", { href: '#' },
                React.createElement(NotchedTag, { type: 'filled', size: 'small', tagStyle: 'style-three' }, "LI"),
                React.createElement("span", null,
                    React.createElement("span", null, "Bar"),
                    React.createElement("span", { style: { fontWeight: 400 } }, "\u00A0(1234)"))))));
};
