// lucid-ui/dist/esm/components/NotchedTag/NotchedTag.stories.js
import React from 'react';
import createClass from 'create-react-class';
import NotchedTag from './NotchedTag';
export default {
    title: 'Visual Design/NotchedTag',
    component: NotchedTag,
    parameters: {
        docs: {
            description: {
                component: NotchedTag.peek.description,
            },
        },
    },
};
/* Basic */
export const Basic = () => {
    const style = { display: 'inline-block', marginRight: '10px' };
    const Component = createClass({
        render() {
            return (React.createElement("div", null,
                React.createElement("div", null,
                    React.createElement(NotchedTag, { style: style }, "ADV"),
                    React.createElement(NotchedTag, { size: 'small', style: style }, "ADV"),
                    React.createElement(NotchedTag, { type: 'filled', style: style }, "ADV"),
                    React.createElement(NotchedTag, { size: 'small', type: 'filled', style: style }, "ADV")),
                React.createElement("div", null,
                    React.createElement(NotchedTag, { tagStyle: 'style-two', style: style }, "IO"),
                    React.createElement(NotchedTag, { size: 'small', tagStyle: 'style-two', style: style }, "IO"),
                    React.createElement(NotchedTag, { tagStyle: 'style-two', type: 'filled', style: style }, "IO"),
                    React.createElement(NotchedTag, { size: 'small', tagStyle: 'style-two', type: 'filled', style: style }, "IO")),
                React.createElement("div", null,
                    React.createElement(NotchedTag, { tagStyle: 'style-three', style: style }, "LI"),
                    React.createElement(NotchedTag, { size: 'small', tagStyle: 'style-three', style: style }, "LI"),
                    React.createElement(NotchedTag, { tagStyle: 'style-three', type: 'filled', style: style }, "LI"),
                    React.createElement(NotchedTag, { size: 'small', tagStyle: 'style-three', type: 'filled', style: style }, "LI"))));
        },
    });
    return React.createElement(Component, null);
};
