// lucid-ui/dist/cjs/components/Line/Line.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Basic = void 0;
var react_1 = __importDefault(require("react"));
var chartConstants = __importStar(require("../../constants/charts"));
var Line_1 = __importDefault(require("./Line"));
exports.default = {
    title: 'Visualizations/Line',
    component: Line_1.default,
    parameters: {
        docs: {
            description: {
                component: Line_1.default.peek.description,
            },
        },
    },
};
/* Basic */
var Basic = function (args) {
    return (react_1.default.createElement("div", null,
        react_1.default.createElement("svg", { width: 200, height: 120 },
            react_1.default.createElement(Line_1.default, __assign({}, args, { color: chartConstants.COLOR_0, d: 'M0,0 L120,120' })),
            react_1.default.createElement(Line_1.default, __assign({}, args, { color: chartConstants.COLOR_1, d: 'M0,20 L120,100' })),
            react_1.default.createElement(Line_1.default, __assign({}, args, { color: chartConstants.COLOR_2, d: 'M0,40 L120,80' })),
            react_1.default.createElement(Line_1.default, __assign({}, args, { color: chartConstants.COLOR_3, d: 'M0,60 L120,60' })),
            react_1.default.createElement(Line_1.default, __assign({}, args, { color: chartConstants.COLOR_4, d: 'M0,80 L120,40' })),
            react_1.default.createElement(Line_1.default, __assign({}, args, { color: chartConstants.COLOR_5, d: 'M0,100 L120,20' })),
            react_1.default.createElement(Line_1.default, __assign({}, args, { color: chartConstants.COLOR_6, d: 'M0,120 L120,0' }))),
        react_1.default.createElement("svg", { width: 200, height: 120 },
            react_1.default.createElement(Line_1.default, __assign({}, args, { isDotted: true, color: chartConstants.COLOR_0, d: 'M0,0 L120,120' })),
            react_1.default.createElement(Line_1.default, __assign({}, args, { isDotted: true, color: chartConstants.COLOR_1, d: 'M0,20 L120,100' })),
            react_1.default.createElement(Line_1.default, __assign({}, args, { isDotted: true, color: chartConstants.COLOR_2, d: 'M0,40 L120,80' })),
            react_1.default.createElement(Line_1.default, __assign({}, args, { isDotted: true, color: chartConstants.COLOR_3, d: 'M0,60 L120,60' })),
            react_1.default.createElement(Line_1.default, __assign({}, args, { isDotted: true, color: chartConstants.COLOR_4, d: 'M0,80 L120,40' })),
            react_1.default.createElement(Line_1.default, __assign({}, args, { isDotted: true, color: chartConstants.COLOR_5, d: 'M0,100 L120,20' })),
            react_1.default.createElement(Line_1.default, __assign({}, args, { isDotted: true, color: chartConstants.COLOR_6, d: 'M0,120 L120,0' })))));
};
exports.Basic = Basic;
