// lucid-ui/dist/cjs/components/EmptyStateWrapper/EmptyStateWrapper.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.YouHaveNoData = exports.IsLoading = exports.Basic = void 0;
var react_1 = __importDefault(require("react"));
var EmptyStateWrapper_1 = require("./EmptyStateWrapper");
exports.default = {
    title: 'Utility/EmptyStateWrapper',
    component: EmptyStateWrapper_1.EmptyStateWrapper,
    parameters: {
        docs: {
            description: {
                component: EmptyStateWrapper_1.EmptyStateWrapper.peek.description,
            },
        },
    },
    argTypes: {
        children: {
            control: false,
        },
    },
};
/* Basic */
var Basic = function (args) {
    return react_1.default.createElement(EmptyStateWrapper_1.EmptyStateWrapper, __assign({}, args));
};
exports.Basic = Basic;
exports.Basic.args = {
    Title: 'Empty State Wrapper',
    isEmpty: true,
    isLoading: false,
    anchorMessage: false,
};
/* isLoading */
var IsLoading = function (args) {
    return react_1.default.createElement(EmptyStateWrapper_1.EmptyStateWrapper, __assign({}, args));
};
exports.IsLoading = IsLoading;
exports.IsLoading.args = {
    isEmpty: false,
    isLoading: true,
    anchorMessage: false,
};
/* YouHaveNoData */
var YouHaveNoData = function (args) {
    return react_1.default.createElement(EmptyStateWrapper_1.EmptyStateWrapper, __assign({}, args));
};
exports.YouHaveNoData = YouHaveNoData;
exports.YouHaveNoData.args = {
    isEmpty: true,
    isLoading: false,
    anchorMessage: false,
};
