// lucid-ui/dist/cjs/components/Validation/Validation.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Basic = void 0;
var react_1 = __importDefault(require("react"));
var Validation_1 = __importDefault(require("./Validation"));
var TextField_1 = __importDefault(require("../TextField/TextField"));
exports.default = {
    title: 'Helpers/Validation',
    component: Validation_1.default,
    parameters: {
        docs: {
            description: {
                component: Validation_1.default.peek.description,
            },
        },
    },
};
/* Basic */
var Basic = function (args) {
    return (react_1.default.createElement("section", null,
        react_1.default.createElement("p", null, "Text field with Error prop (Method 1)"),
        react_1.default.createElement(Validation_1.default, __assign({}, args, { Error: 'Error One' }),
            react_1.default.createElement(TextField_1.default, { value: 'Text Field Text' })),
        react_1.default.createElement("br", null),
        react_1.default.createElement("p", null, "Text field with Validation Error child component (Method 2)"),
        react_1.default.createElement(Validation_1.default, __assign({}, args),
            react_1.default.createElement(Validation_1.default.Error, null,
                react_1.default.createElement("span", null, "Error Two")),
            react_1.default.createElement(TextField_1.default, { value: 'Text Field Text' })),
        react_1.default.createElement("br", null),
        react_1.default.createElement("p", null, "Text area with Error prop (Method 1)"),
        react_1.default.createElement(Validation_1.default, __assign({}, args, { Error: 'Error Three' }),
            react_1.default.createElement(TextField_1.default, { isMultiLine: true, rows: 3, value: 'Text Area Text' })),
        react_1.default.createElement("br", null),
        react_1.default.createElement("p", null, "Text area with Validation Error child component (Method 2)"),
        react_1.default.createElement(Validation_1.default, __assign({}, args, { Error: 'Error FOUR TEST' }),
            react_1.default.createElement(Validation_1.default.Error, null,
                react_1.default.createElement("span", null, "Error Four")),
            react_1.default.createElement(TextField_1.default, { isMultiLine: true, rows: 3, value: 'Text Area Text' }))));
};
exports.Basic = Basic;
