// lucid-ui/dist/cjs/components/Bar/Bar.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RegularColors = exports.CustomDimensionUnits = exports.CustomColors = void 0;
var lodash_1 = __importDefault(require("lodash"));
var react_1 = __importDefault(require("react"));
var chartConstants = __importStar(require("../../constants/charts"));
var Bar_1 = __importDefault(require("../Bar/Bar"));
exports.default = {
    title: 'Visualizations/Bar',
    component: Bar_1.default,
    parameters: {
        docs: {
            description: {
                component: Bar_1.default.peek.description,
            },
        },
    },
};
/* Custom Colors */
var CustomColors = function (args) {
    var svgProps = {
        width: 20,
        height: 100,
    };
    var pointProps = {
        x: 5,
        y: 0,
        width: 10,
        height: 100,
    };
    return (react_1.default.createElement("div", null,
        react_1.default.createElement("svg", __assign({}, svgProps),
            react_1.default.createElement(Bar_1.default, __assign({}, args, pointProps, { color: '#f80' }))),
        react_1.default.createElement("svg", __assign({}, svgProps),
            react_1.default.createElement(Bar_1.default, __assign({}, args, pointProps, { color: '#abc123' })))));
};
exports.CustomColors = CustomColors;
exports.CustomColors.storyName = 'CustomColors';
/* Custom Dimension Units */
var CustomDimensionUnits = function (args) {
    var svgProps = {
        width: 20,
        height: 100,
    };
    var pointProps = {
        x: 5,
        y: 0,
        width: '10%',
        height: '100%',
    };
    return (react_1.default.createElement("div", null,
        react_1.default.createElement("svg", __assign({}, svgProps),
            react_1.default.createElement(Bar_1.default, __assign({}, args, pointProps, { color: '#f80' }))),
        react_1.default.createElement("svg", __assign({}, svgProps),
            react_1.default.createElement(Bar_1.default, __assign({}, args, pointProps, { color: '#abc123' })))));
};
exports.CustomDimensionUnits = CustomDimensionUnits;
/* Regular Colors */
var RegularColors = function (args) {
    var svgProps = {
        width: 20,
        height: 100,
    };
    var barProps = {
        x: 5,
        y: 0,
        width: 10,
        height: 100,
    };
    var colors = [
        chartConstants.COLOR_0_LIGHTEST,
        chartConstants.COLOR_0_LIGHT,
        chartConstants.COLOR_0,
        chartConstants.COLOR_0_DARK,
        chartConstants.COLOR_0_DARKEST,
        chartConstants.COLOR_1_LIGHTEST,
        chartConstants.COLOR_1_LIGHT,
        chartConstants.COLOR_1,
        chartConstants.COLOR_1_DARK,
        chartConstants.COLOR_1_DARKEST,
        chartConstants.COLOR_2_LIGHTEST,
        chartConstants.COLOR_2_LIGHT,
        chartConstants.COLOR_2,
        chartConstants.COLOR_2_DARK,
        chartConstants.COLOR_2_DARKEST,
        chartConstants.COLOR_3_LIGHTEST,
        chartConstants.COLOR_3_LIGHT,
        chartConstants.COLOR_3,
        chartConstants.COLOR_3_DARK,
        chartConstants.COLOR_3_DARKEST,
        chartConstants.COLOR_4_LIGHTEST,
        chartConstants.COLOR_4_LIGHT,
        chartConstants.COLOR_4,
        chartConstants.COLOR_4_DARK,
        chartConstants.COLOR_4_DARKEST,
        chartConstants.COLOR_5_LIGHTEST,
        chartConstants.COLOR_5_LIGHT,
        chartConstants.COLOR_5,
        chartConstants.COLOR_5_DARK,
        chartConstants.COLOR_5_DARKEST,
        chartConstants.COLOR_6_LIGHTEST,
        chartConstants.COLOR_6_LIGHT,
        chartConstants.COLOR_6,
        chartConstants.COLOR_6_DARK,
        chartConstants.COLOR_6_DARKEST,
        chartConstants.COLOR_GOOD_LIGHT,
        chartConstants.COLOR_GOOD,
        chartConstants.COLOR_GOOD_DARK,
        chartConstants.COLOR_BAD_LIGHT,
        chartConstants.COLOR_BAD,
        chartConstants.COLOR_BAD_DARK,
        chartConstants.COLOR_NEUTRAL,
    ];
    return (react_1.default.createElement("div", null, lodash_1.default.map(colors, function (color) { return (react_1.default.createElement("svg", __assign({ key: color }, svgProps),
        react_1.default.createElement(Bar_1.default, __assign({}, args, barProps, { color: color })))); })));
};
exports.RegularColors = RegularColors;
