// lucid-ui/dist/cjs/components/SlidePanel/SlidePanel.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoopedSlides = void 0;
var react_1 = __importStar(require("react"));
var SlidePanel_1 = __importDefault(require("./SlidePanel"));
var Button_1 = __importDefault(require("../Button/Button"));
var AnalyzeDataIcon_1 = __importDefault(require("../Icon/AnalyzeDataIcon/AnalyzeDataIcon"));
var CalendarIcon_1 = __importDefault(require("../Icon/CalendarIcon/CalendarIcon"));
var DuplicateIcon_1 = __importDefault(require("../Icon/DuplicateIcon/DuplicateIcon"));
var EditIcon_1 = __importDefault(require("../Icon/EditIcon/EditIcon"));
var FileIcon_1 = __importDefault(require("../Icon/FileIcon/FileIcon"));
var ImageIcon_1 = __importDefault(require("../Icon/ImageIcon/ImageIcon"));
var SettingsIcon_1 = __importDefault(require("../Icon/SettingsIcon/SettingsIcon"));
exports.default = {
    title: 'Private/SlidePanel',
    component: SlidePanel_1.default,
    parameters: {
        docs: {
            description: {
                component: SlidePanel_1.default.peek.description,
            },
        },
    },
    args: SlidePanel_1.default.defaultProps,
};
/* Looped Slides */
var LoopedSlides = function (args) {
    var Slide = SlidePanel_1.default.Slide;
    var _a = (0, react_1.useState)(0), offset = _a[0], setOffset = _a[1];
    var handlePrev = function () {
        setOffset(offset - 1);
    };
    var handleNext = function () {
        setOffset(offset + 1);
    };
    var handleSwipe = function (slidesSwiped) {
        setOffset(offset + slidesSwiped);
    };
    return (react_1.default.createElement("section", null,
        react_1.default.createElement(Button_1.default, { onClick: handlePrev }, "Backward"),
        react_1.default.createElement(Button_1.default, { onClick: handleNext }, "Forward"),
        "Current offset: ",
        offset,
        react_1.default.createElement(SlidePanel_1.default, __assign({}, args, { slidesToShow: 2, offset: offset, onSwipe: handleSwipe, isLooped: true }),
            react_1.default.createElement(Slide, null,
                react_1.default.createElement(AnalyzeDataIcon_1.default, { style: {
                        width: '100%',
                        height: '30vh',
                        background: 'whitesmoke',
                    } })),
            react_1.default.createElement(Slide, null,
                react_1.default.createElement(CalendarIcon_1.default, { style: { width: '100%', height: '30vh' } })),
            react_1.default.createElement(Slide, null,
                react_1.default.createElement(DuplicateIcon_1.default, { style: {
                        width: '100%',
                        height: '30vh',
                        background: 'whitesmoke',
                    } })),
            react_1.default.createElement(Slide, null,
                react_1.default.createElement(EditIcon_1.default, { style: { width: '100%', height: '30vh' } })),
            react_1.default.createElement(Slide, null,
                react_1.default.createElement(FileIcon_1.default, { style: {
                        width: '100%',
                        height: '30vh',
                        background: 'whitesmoke',
                    } })),
            react_1.default.createElement(Slide, null,
                react_1.default.createElement(ImageIcon_1.default, { style: {
                        width: '100%',
                        height: '30vh',
                        background: 'whitesmoke',
                    } })),
            react_1.default.createElement(Slide, null,
                react_1.default.createElement(SettingsIcon_1.default, { style: { width: '100%', height: '30vh' } })))));
};
exports.LoopedSlides = LoopedSlides;
