// lucid-ui/dist/cjs/components/Lines/Lines.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomColors = exports.Basic = void 0;
var lodash_1 = __importDefault(require("lodash"));
var react_1 = __importDefault(require("react"));
var Lines_1 = __importDefault(require("./Lines"));
var chartConstants = __importStar(require("../../constants/charts"));
var d3Scale = __importStar(require("d3-scale"));
exports.default = {
    title: 'Visualizations/Lines',
    component: Lines_1.default,
    parameters: {
        docs: {
            description: {
                component: Lines_1.default.peek.description,
            },
        },
    },
};
/* Basic */
var Basic = function (args) {
    var width = 800;
    var height = 400;
    var data = [
        { x: 'one', y0: 20, y1: 18, y2: 16, y3: 14, y4: 12, y5: 10, y6: 8 },
        { x: 'two', y0: 2, y1: 3, y2: 4, y3: 6, y4: 8, y5: 9, y6: 6 },
        { x: 'three', y0: 2, y1: 4, y2: 5, y3: 6, y4: 9, y5: 10, y6: 3 },
        { x: 'four', y0: 3, y1: 6, y2: 7, y3: 7, y4: 7, y5: 4, y6: 2 },
        { x: 'five', y0: 4, y1: 8, y2: 9, y3: 8, y4: 5, y5: 4, y6: 3 },
        { x: 'six', y0: 20, y1: 18, y2: 16, y3: 14, y4: 12, y5: 10, y6: 8 },
    ];
    var yFields = ['y0', 'y1', 'y2', 'y3', 'y4', 'y5', 'y6'];
    var yMax = lodash_1.default.max(lodash_1.default.reduce(yFields, function (acc, field) {
        return acc.concat(lodash_1.default.map(data, field));
    }, []));
    var xScale = d3Scale
        .scalePoint()
        .domain(lodash_1.default.map(data, 'x'))
        .range([0, width]);
    var yScale = d3Scale.scaleLinear().domain([0, yMax]).range([height, 0]);
    return (react_1.default.createElement("div", null,
        react_1.default.createElement("svg", { width: width, height: height },
            react_1.default.createElement(Lines_1.default, __assign({}, args, { data: data, xScale: xScale, yScale: yScale, yFields: yFields }))),
        react_1.default.createElement("svg", { width: width, height: height },
            react_1.default.createElement(Lines_1.default, __assign({}, args, { data: data, xScale: xScale, yScale: yScale, yFields: yFields, isStacked: true })))));
};
exports.Basic = Basic;
/* Custom Colors */
var CustomColors = function (args) {
    /* eslint-disable comma-spacing */
    var width = 1000;
    var height = 400;
    var data = [
        { x: 'one', y0: 1, y1: 2, y2: 3, y3: 5 },
        { x: 'two', y0: 2, y1: 3, y2: 4, y3: 6 },
        { x: 'three', y0: 2, y1: 4, y2: 5, y3: 6 },
        { x: 'four', y0: 3, y1: 6, y2: 7, y3: 7 },
        { x: 'five', y0: 4, y1: 8, y2: 9, y3: 8 },
        { x: 'six', y0: 20, y1: 8, y2: 9, y3: 1 },
    ];
    var yFields = ['y0', 'y1', 'y2', 'y3'];
    var yMax = lodash_1.default.max(lodash_1.default.reduce(yFields, function (acc, field) {
        return acc.concat(lodash_1.default.map(data, field));
    }, []));
    var xScale = d3Scale
        .scalePoint()
        .domain(lodash_1.default.map(data, 'x'))
        .range([0, width]);
    var yScale = d3Scale.scaleLinear().domain([0, yMax]).range([height, 0]);
    return (react_1.default.createElement("div", null,
        react_1.default.createElement("svg", { width: width, height: height },
            react_1.default.createElement(Lines_1.default, __assign({}, args, { data: data, xScale: xScale, yScale: yScale, yFields: yFields, colorMap: {
                    y0: chartConstants.COLOR_BAD,
                    y1: chartConstants.COLOR_GOOD,
                    y2: '#ff8800',
                    y3: '#abc123',
                } })))));
};
exports.CustomColors = CustomColors;
