// lucid-ui/dist/cjs/components/ButtonGroup/ButtonGroup.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OnClick = exports.Disabled = exports.Stateless = exports.Stateful = exports.Basic = void 0;
var react_1 = __importStar(require("react"));
var lodash_1 = __importDefault(require("lodash"));
var ButtonGroup_1 = __importDefault(require("./ButtonGroup"));
exports.default = {
    title: 'Controls/ButtonGroup',
    component: ButtonGroup_1.default,
    subcomponents: { 'ButtonGroup.Button': ButtonGroup_1.default.Button },
    args: ButtonGroup_1.default.defaultProps,
    parameters: {
        docs: {
            inlineStories: false,
            description: {
                component: ButtonGroup_1.default.peek.description,
            },
        },
    },
};
var Basic = function (args) {
    var buttonStyle = { width: '100px' };
    return (react_1.default.createElement(ButtonGroup_1.default, __assign({}, args),
        react_1.default.createElement(ButtonGroup_1.default.Button, { style: buttonStyle }, "Smol"),
        react_1.default.createElement(ButtonGroup_1.default.Button, { style: buttonStyle }, "Lonnnnnnnng"),
        react_1.default.createElement(ButtonGroup_1.default.Button, { style: buttonStyle }, "Medium")));
};
exports.Basic = Basic;
var Stateful = function (args) {
    var _a = (0, react_1.useState)([]), selectedIndices = _a[0], setSelectedIndices = _a[1];
    var handleSelect = function (selectedIndex, _a) {
        var event = _a.event, props = _a.props;
        var currentIndices = selectedIndices;
        // If the item does not exist in the original array add it, if it exists remove it.
        setSelectedIndices(lodash_1.default.xor(currentIndices, [selectedIndex]));
    };
    return (react_1.default.createElement(ButtonGroup_1.default, __assign({}, args, { selectedIndices: selectedIndices, onSelect: handleSelect }),
        react_1.default.createElement(ButtonGroup_1.default.Button, null, "Zero"),
        react_1.default.createElement(ButtonGroup_1.default.Button, null, "One"),
        react_1.default.createElement(ButtonGroup_1.default.Button, null, "Two"),
        react_1.default.createElement(ButtonGroup_1.default.Button, null, "Three")));
};
exports.Stateful = Stateful;
var Stateless = function (args) {
    return (react_1.default.createElement(ButtonGroup_1.default, __assign({}, args, { selectedIndices: [7, 8] }),
        react_1.default.createElement(ButtonGroup_1.default.Button, null, "Zero"),
        react_1.default.createElement(ButtonGroup_1.default.Button, null, "One"),
        react_1.default.createElement(ButtonGroup_1.default.Button, null, "Two"),
        react_1.default.createElement(ButtonGroup_1.default.Button, null, "Three"),
        react_1.default.createElement(ButtonGroup_1.default.Button, null, "Four"),
        react_1.default.createElement(ButtonGroup_1.default.Button, { isDisabled: true }, "Five"),
        react_1.default.createElement(ButtonGroup_1.default.Button, null, "Six"),
        react_1.default.createElement(ButtonGroup_1.default.Button, null, "Seven"),
        react_1.default.createElement(ButtonGroup_1.default.Button, null, "Eight"),
        react_1.default.createElement(ButtonGroup_1.default.Button, null, "Nine")));
};
exports.Stateless = Stateless;
var Disabled = function (args) {
    var buttonStyle = { width: '100px' };
    return (react_1.default.createElement(ButtonGroup_1.default, __assign({}, args),
        react_1.default.createElement(ButtonGroup_1.default.Button, { style: buttonStyle }, "Zero"),
        react_1.default.createElement(ButtonGroup_1.default.Button, { isDisabled: true, style: buttonStyle }, "One"),
        react_1.default.createElement(ButtonGroup_1.default.Button, { style: buttonStyle }, "Two"),
        react_1.default.createElement(ButtonGroup_1.default.Button, { style: buttonStyle }, "Three")));
};
exports.Disabled = Disabled;
var OnClick = function (args) {
    var handleClick = function (label) {
        alert(label);
    };
    return (react_1.default.createElement(ButtonGroup_1.default, __assign({}, args),
        react_1.default.createElement(ButtonGroup_1.default.Button, { onClick: function () { return handleClick('Zero'); } }, "Zero"),
        react_1.default.createElement(ButtonGroup_1.default.Button, { onClick: function () { return handleClick('One'); } }, "One")));
};
exports.OnClick = OnClick;
