// lucid-ui/dist/cjs/components/InfiniteSlidePanel/InfiniteSlidePanel.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DynamicContent = void 0;
var react_1 = __importStar(require("react"));
var InfiniteSlidePanel_1 = __importDefault(require("./InfiniteSlidePanel"));
var Button_1 = __importDefault(require("../Button/Button"));
exports.default = {
    title: 'Private/InfiniteSlidePanel',
    component: InfiniteSlidePanel_1.default,
    parameters: {
        docs: {
            description: {
                component: InfiniteSlidePanel_1.default.peek.description,
            },
        },
    },
    args: InfiniteSlidePanel_1.default.defaultProps,
};
/* Dynamic Content */
var DynamicContent = function (args) {
    var generateRGB = function (n) {
        var R = Math.floor((Math.sin(n / Math.PI) + 1) * 128);
        var G = Math.floor((Math.sin((2 * n) / Math.PI) + 1) * 128);
        var B = Math.floor((Math.sin((3 * n) / Math.PI) + 1) * 128);
        return "rgb(".concat(R, ",").concat(G, ",").concat(B, ")");
    };
    var _a = (0, react_1.useState)(0), offset = _a[0], setOffset = _a[1];
    var handlePrev = function () {
        setOffset(offset - 1);
    };
    var handleNext = function () {
        setOffset(offset + 1);
    };
    var handleSwipe = function (slidesSwiped) {
        setOffset(offset + slidesSwiped);
    };
    return (react_1.default.createElement("section", null,
        react_1.default.createElement(Button_1.default, { onClick: handlePrev }, "Backward"),
        react_1.default.createElement(Button_1.default, { onClick: handleNext }, "Forward"),
        react_1.default.createElement("span", { style: { marginLeft: 10 } },
            "Current offset: ",
            offset),
        react_1.default.createElement(InfiniteSlidePanel_1.default, __assign({}, args, { totalSlides: 12, slidesToShow: 3, offset: offset, onSwipe: handleSwipe }), function (slideOffset) { return (react_1.default.createElement("div", { style: {
                width: '100%',
                height: '30vh',
                background: generateRGB(slideOffset),
                display: 'flex',
                justifyContent: 'center',
                alignItems: 'center',
            } }, slideOffset)); })));
};
exports.DynamicContent = DynamicContent;
