// lucid-ui/dist/esm/components/Checkbox/Checkbox.stories.js
import React, { useState } from 'react';
import Checkbox from './Checkbox';
export default {
    title: 'Controls/Checkbox',
    component: Checkbox,
    parameters: {
        docs: {
            description: {
                component: Checkbox.peek.description,
            },
        },
    },
};
const Template = (args) => {
    const [selected, setSelected] = useState(args.isSelected || false);
    return (React.createElement(Checkbox, { ...args, isSelected: selected, onSelect: () => setSelected(!selected) }));
};
export const Basic = Template.bind({});
Basic.args = {
    title: 'Default',
};
export const Plain = Template.bind({});
Plain.args = {
    title: 'Plain',
    isDisabled: false,
    isSelected: true,
};
export const DisabledUnselected = Template.bind({});
DisabledUnselected.args = {
    title: 'Disabled Unselected',
    isDisabled: true,
    isSelected: false,
};
export const DisabledSelected = Template.bind({});
DisabledSelected.args = {
    title: 'Disabled Selected',
    isDisabled: true,
    isSelected: true,
};
export const DisabledIndeterminate = Template.bind({});
DisabledIndeterminate.args = {
    title: 'Disabled Selected',
    isIndeterminate: true,
    isDisabled: true,
};
