// lucid-ui/dist/esm/components/EmptyStateWrapper/EmptyStateWrapper.stories.js
import React from 'react';
import { EmptyStateWrapper, } from './EmptyStateWrapper';
export default {
    title: 'Utility/EmptyStateWrapper',
    component: EmptyStateWrapper,
    parameters: {
        docs: {
            description: {
                component: EmptyStateWrapper.peek.description,
            },
        },
    },
    argTypes: {
        children: {
            control: false,
        },
    },
};
/* Basic */
export const Basic = (args) => {
    return React.createElement(EmptyStateWrapper, { ...args });
};
Basic.args = {
    Title: 'Empty State Wrapper',
    isEmpty: true,
    isLoading: false,
    anchorMessage: false,
};
/* isLoading */
export const IsLoading = (args) => {
    return React.createElement(EmptyStateWrapper, { ...args });
};
IsLoading.args = {
    isEmpty: false,
    isLoading: true,
    anchorMessage: false,
};
/* YouHaveNoData */
export const YouHaveNoData = (args) => {
    return React.createElement(EmptyStateWrapper, { ...args });
};
YouHaveNoData.args = {
    isEmpty: true,
    isLoading: false,
    anchorMessage: false,
};
