// lucid-ui/dist/cjs/components/Checkbox/Checkbox.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DisabledIndeterminate = exports.DisabledSelected = exports.DisabledUnselected = exports.Plain = exports.Basic = void 0;
var react_1 = __importStar(require("react"));
var Checkbox_1 = __importDefault(require("./Checkbox"));
exports.default = {
    title: 'Controls/Checkbox',
    component: Checkbox_1.default,
    parameters: {
        docs: {
            description: {
                component: Checkbox_1.default.peek.description,
            },
        },
    },
};
var Template = function (args) {
    var _a = (0, react_1.useState)(args.isSelected || false), selected = _a[0], setSelected = _a[1];
    return (react_1.default.createElement(Checkbox_1.default, __assign({}, args, { isSelected: selected, onSelect: function () { return setSelected(!selected); } })));
};
exports.Basic = Template.bind({});
exports.Basic.args = {
    title: 'Default',
};
exports.Plain = Template.bind({});
exports.Plain.args = {
    title: 'Plain',
    isDisabled: false,
    isSelected: true,
};
exports.DisabledUnselected = Template.bind({});
exports.DisabledUnselected.args = {
    title: 'Disabled Unselected',
    isDisabled: true,
    isSelected: false,
};
exports.DisabledSelected = Template.bind({});
exports.DisabledSelected.args = {
    title: 'Disabled Selected',
    isDisabled: true,
    isSelected: true,
};
exports.DisabledIndeterminate = Template.bind({});
exports.DisabledIndeterminate.args = {
    title: 'Disabled Selected',
    isIndeterminate: true,
    isDisabled: true,
};
