// lucid-ui/dist/cjs/components/LoadingSkeletons/ComplexTableLoadingSkeleton.stories.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ThreeRowsOneColumns = exports.AddHeader = exports.Basic = void 0;
var react_1 = __importDefault(require("react"));
var create_react_class_1 = __importDefault(require("create-react-class"));
var ComplexTableLoadingSkeleton_1 = __importDefault(require("./ComplexTableLoadingSkeleton"));
exports.default = {
    title: 'Loading Indicator/ComplexTableLoadingSkeleton',
    component: ComplexTableLoadingSkeleton_1.default,
    parameters: {
        docs: {
            description: {
                component: ComplexTableLoadingSkeleton_1.default.peek.description,
            },
        },
    },
};
/* Basic */
var Basic = function () {
    var Component = (0, create_react_class_1.default)({
        render: function () {
            return react_1.default.createElement(ComplexTableLoadingSkeleton_1.default, { isLoading: true });
        },
    });
    return react_1.default.createElement(Component, null);
};
exports.Basic = Basic;
/* Add Header */
var AddHeader = function () {
    var Component = (0, create_react_class_1.default)({
        render: function () {
            return (react_1.default.createElement(ComplexTableLoadingSkeleton_1.default, { isLoading: true, header: 'Added Header' }));
        },
    });
    return react_1.default.createElement(Component, null);
};
exports.AddHeader = AddHeader;
exports.AddHeader.storyName = 'AddHeader';
/* Three Rows One Columns */
var ThreeRowsOneColumns = function () {
    var Component = (0, create_react_class_1.default)({
        render: function () {
            return (react_1.default.createElement("div", null,
                react_1.default.createElement(ComplexTableLoadingSkeleton_1.default, { isLoading: true, width: 860, numRows: 3 })));
        },
    });
    return react_1.default.createElement(Component, null);
};
exports.ThreeRowsOneColumns = ThreeRowsOneColumns;
exports.ThreeRowsOneColumns.storyName = 'ThreeRowsOneColumns';
