// lucid-ui/dist/cjs/components/LoadingSkeletons/HeaderLoadingSkeleton.stories.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Basic = void 0;
var react_1 = __importDefault(require("react"));
var create_react_class_1 = __importDefault(require("create-react-class"));
var HeaderLoadingSkeleton_1 = __importDefault(require("./HeaderLoadingSkeleton"));
exports.default = {
    title: 'Loading Indicator/HeaderLoadingSkeleton',
    component: HeaderLoadingSkeleton_1.default,
    parameters: {
        docs: {
            description: {
                component: HeaderLoadingSkeleton_1.default.peek.description,
            },
        },
    },
};
/* Basic */
var Basic = function () {
    var Component = (0, create_react_class_1.default)({
        render: function () {
            return react_1.default.createElement(HeaderLoadingSkeleton_1.default, { isLoading: true });
        },
    });
    return react_1.default.createElement(Component, null);
};
exports.Basic = Basic;
