// lucid-ui/dist/cjs/components/LoadingSkeletons/LineChartLoadingSkeleton.stories.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TwoRowsTwoColumns = exports.AddHeader = exports.Basic = void 0;
var react_1 = __importDefault(require("react"));
var create_react_class_1 = __importDefault(require("create-react-class"));
var LineChartLoadingSkeleton_1 = __importDefault(require("./LineChartLoadingSkeleton"));
exports.default = {
    title: 'Loading Indicator/LineChartLoadingSkeleton',
    component: LineChartLoadingSkeleton_1.default,
    parameters: {
        docs: {
            description: {
                component: LineChartLoadingSkeleton_1.default.peek.description,
            },
        },
    },
};
/* Basic */
var Basic = function () {
    var Component = (0, create_react_class_1.default)({
        render: function () {
            return react_1.default.createElement(LineChartLoadingSkeleton_1.default, { isLoading: true });
        },
    });
    return react_1.default.createElement(Component, null);
};
exports.Basic = Basic;
/* Add Header */
var AddHeader = function () {
    var Component = (0, create_react_class_1.default)({
        render: function () {
            return (react_1.default.createElement(LineChartLoadingSkeleton_1.default, { isLoading: true, header: 'Added Header' }));
        },
    });
    return react_1.default.createElement(Component, null);
};
exports.AddHeader = AddHeader;
exports.AddHeader.storyName = 'AddHeader';
/* Two Rows Two Columns */
var TwoRowsTwoColumns = function () {
    var Component = (0, create_react_class_1.default)({
        render: function () {
            return (react_1.default.createElement("div", null,
                react_1.default.createElement(LineChartLoadingSkeleton_1.default, { isLoading: true, width: 400, numRows: 2, numColumns: 2, marginRight: 50, marginBottom: 50 })));
        },
    });
    return react_1.default.createElement(Component, null);
};
exports.TwoRowsTwoColumns = TwoRowsTwoColumns;
exports.TwoRowsTwoColumns.storyName = 'TwoRowsTwoColumns';
