// lucid-ui/dist/cjs/components/LoadingSkeletons/SimpleTableLoadingSkeleton.stories.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ThreeRowsOneColumn = exports.AddHeader = exports.Basic = void 0;
var react_1 = __importDefault(require("react"));
var create_react_class_1 = __importDefault(require("create-react-class"));
var SimpleTableLoadingSkeleton_1 = __importDefault(require("./SimpleTableLoadingSkeleton"));
exports.default = {
    title: 'Loading Indicator/SimpleTableLoadingSkeleton',
    component: SimpleTableLoadingSkeleton_1.default,
    parameters: {
        docs: {
            description: {
                component: SimpleTableLoadingSkeleton_1.default.peek.description,
            },
        },
    },
};
/* Basic */
var Basic = function () {
    var Component = (0, create_react_class_1.default)({
        render: function () {
            return react_1.default.createElement(SimpleTableLoadingSkeleton_1.default, { isLoading: true });
        },
    });
    return react_1.default.createElement(Component, null);
};
exports.Basic = Basic;
/* Add Header */
var AddHeader = function () {
    var Component = (0, create_react_class_1.default)({
        render: function () {
            return (react_1.default.createElement(SimpleTableLoadingSkeleton_1.default, { isLoading: true, height: 50, header: 'Added Header' }));
        },
    });
    return react_1.default.createElement(Component, null);
};
exports.AddHeader = AddHeader;
exports.AddHeader.storyName = 'AddHeader';
/* Three Rows One Column */
var ThreeRowsOneColumn = function () {
    var Component = (0, create_react_class_1.default)({
        render: function () {
            return (react_1.default.createElement("div", null,
                react_1.default.createElement(SimpleTableLoadingSkeleton_1.default, { isLoading: true, height: 70, numRows: 3 })));
        },
    });
    return react_1.default.createElement(Component, null);
};
exports.ThreeRowsOneColumn = ThreeRowsOneColumn;
exports.ThreeRowsOneColumn.storyName = 'ThreeRowsOneColumn';
