// lucid-ui/dist/cjs/components/LoadingSkeletons/SingleLineLoadingSkeleton.stories.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TwoRowsThreeColumns = exports.AddHeader = exports.Basic = void 0;
var react_1 = __importDefault(require("react"));
var create_react_class_1 = __importDefault(require("create-react-class"));
var SingleLineLoadingSkeleton_1 = __importDefault(require("./SingleLineLoadingSkeleton"));
exports.default = {
    title: 'Loading Indicator/SingleLineLoadingSkeleton',
    component: SingleLineLoadingSkeleton_1.default,
    parameters: {
        docs: {
            description: {
                component: SingleLineLoadingSkeleton_1.default.peek.description,
            },
        },
    },
};
/* Basic */
var Basic = function () {
    var Component = (0, create_react_class_1.default)({
        render: function () {
            return react_1.default.createElement(SingleLineLoadingSkeleton_1.default, { isLoading: true });
        },
    });
    return react_1.default.createElement(Component, null);
};
exports.Basic = Basic;
/* Add Header */
var AddHeader = function () {
    var Component = (0, create_react_class_1.default)({
        render: function () {
            return (react_1.default.createElement(SingleLineLoadingSkeleton_1.default, { isLoading: true, width: 700, height: 50, header: 'Added Header' }));
        },
    });
    return react_1.default.createElement(Component, null);
};
exports.AddHeader = AddHeader;
exports.AddHeader.storyName = 'AddHeader';
/* Two Rows Three Columns */
var TwoRowsThreeColumns = function () {
    var Component = (0, create_react_class_1.default)({
        render: function () {
            return (react_1.default.createElement("div", null,
                react_1.default.createElement(SingleLineLoadingSkeleton_1.default, { isLoading: true, width: 250, height: 20, numRows: 2, numColumns: 3, marginRight: 20, marginBottom: 10 })));
        },
    });
    return react_1.default.createElement(Component, null);
};
exports.TwoRowsThreeColumns = TwoRowsThreeColumns;
exports.TwoRowsThreeColumns.storyName = 'TwoRowsThreeColumns';
