// lucid-ui/dist/cjs/components/TimeSelect/TimeSelect.stories.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TwentyFourHour = exports.TwelveHourDisabled = exports.TwelveHour = void 0;
var react_1 = __importStar(require("react"));
var TimeSelect_1 = __importDefault(require("./TimeSelect"));
exports.default = {
    title: 'Controls/TimeSelect',
    component: TimeSelect_1.default,
    parameters: {
        docs: {
            description: {
                component: TimeSelect_1.default.peek.description,
            },
        },
    },
};
/* Twelve Hour */
var TwelveHour = function () {
    var _a = (0, react_1.useState)(new Date()), time = _a[0], setTime = _a[1];
    return react_1.default.createElement(TimeSelect_1.default, { time: time, onChange: setTime });
};
exports.TwelveHour = TwelveHour;
exports.TwelveHour.storyName = 'TwelveHour';
/* Twelve Hour Disabled */
var TwelveHourDisabled = function () {
    var _a = (0, react_1.useState)(new Date()), time = _a[0], setTime = _a[1];
    return react_1.default.createElement(TimeSelect_1.default, { isDisabled: true, time: time, onChange: setTime });
};
exports.TwelveHourDisabled = TwelveHourDisabled;
exports.TwelveHourDisabled.storyName = 'TwelveHourDisabled';
/* Twenty Four Hour */
var TwentyFourHour = function () {
    var _a = (0, react_1.useState)(new Date()), time = _a[0], setTime = _a[1];
    return react_1.default.createElement(TimeSelect_1.default, { time: time, is24HourClock: true, onChange: setTime });
};
exports.TwentyFourHour = TwentyFourHour;
exports.TwentyFourHour.storyName = 'TwentyFourHour';
