// lucid-ui/dist/cjs/components/Breadcrumb/Breadcrumb.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Rich = exports.Basic = void 0;
var react_1 = __importDefault(require("react"));
var Breadcrumb_1 = __importDefault(require("./Breadcrumb"));
var NotchedTag_1 = __importDefault(require("../NotchedTag/NotchedTag"));
exports.default = {
    title: 'Navigation/Breadcrumb',
    component: Breadcrumb_1.default,
    parameters: {
        docs: {
            description: {
                component: Breadcrumb_1.default.peek.description,
            },
        },
    },
};
/* Basic */
var Basic = function (args) {
    var Item = Breadcrumb_1.default.Item;
    return (react_1.default.createElement(Breadcrumb_1.default, __assign({}, args),
        react_1.default.createElement(Item, { key: 'home' },
            react_1.default.createElement("a", { href: '#' }, "Start")),
        react_1.default.createElement(Item, { key: 'list' },
            react_1.default.createElement("a", { href: '#' }, "List")),
        react_1.default.createElement(Item, { key: 'item' },
            react_1.default.createElement("a", { href: '#' }, "Item"))));
};
exports.Basic = Basic;
/* Rich */
var Rich = function (args) {
    var Item = Breadcrumb_1.default.Item;
    return (react_1.default.createElement(Breadcrumb_1.default, __assign({}, args),
        react_1.default.createElement(Item, { key: 'home' },
            react_1.default.createElement("a", { href: '#' },
                react_1.default.createElement(NotchedTag_1.default, { type: 'stroke', size: 'small', tagStyle: 'style-two' }, "IO"),
                react_1.default.createElement("span", null,
                    react_1.default.createElement("span", null, "Foo"),
                    react_1.default.createElement("span", { style: { fontWeight: 400 } }, "\u00A0(1234)")))),
        react_1.default.createElement(Item, { key: 'list' },
            react_1.default.createElement("a", { href: '#' },
                react_1.default.createElement(NotchedTag_1.default, { type: 'filled', size: 'small', tagStyle: 'style-three' }, "LI"),
                react_1.default.createElement("span", null,
                    react_1.default.createElement("span", null, "Bar"),
                    react_1.default.createElement("span", { style: { fontWeight: 400 } }, "\u00A0(1234)"))))));
};
exports.Rich = Rich;
