// lucid-ui/dist/cjs/components/PieChart/PieChart.stories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SmallWithNoStrokeOrHover = exports.Percents = exports.ColorMap = exports.LegendTest = exports.BasicDonuts = exports.Basic = void 0;
var lodash_1 = __importDefault(require("lodash"));
var react_1 = __importDefault(require("react"));
var chartConstants = __importStar(require("../../constants/charts"));
var PieChart_1 = __importDefault(require("./PieChart"));
var Legend_1 = __importDefault(require("../Legend/Legend"));
exports.default = {
    title: 'Visualizations/PieChart',
    component: PieChart_1.default,
    parameters: {
        docs: {
            description: {
                component: PieChart_1.default.peek.description,
            },
        },
    },
    args: PieChart_1.default.defaultProps,
};
/* Basic */
var Basic = function (args) {
    var data = [
        { x: 'Leslie', y: 60 },
        { x: 'Ron', y: 40 },
        { x: 'Tom', y: 30 },
        { x: 'Gary', y: 20 },
        { x: 'Ben', y: 15 },
    ];
    var style = {
        paddingTop: '4rem',
    };
    return (react_1.default.createElement("section", { style: style },
        react_1.default.createElement(PieChart_1.default, { data: data }),
        react_1.default.createElement(PieChart_1.default, __assign({ data: data }, args, { palette: chartConstants.PALETTE_MONOCHROME_0_5 })),
        react_1.default.createElement(PieChart_1.default, __assign({}, args, { data: data, palette: chartConstants.PALETTE_MONOCHROME_1_5 })),
        react_1.default.createElement(PieChart_1.default, __assign({}, args, { data: data, palette: chartConstants.PALETTE_MONOCHROME_2_5 })),
        react_1.default.createElement(PieChart_1.default, __assign({}, args, { data: data, palette: chartConstants.PALETTE_MONOCHROME_3_5 })),
        react_1.default.createElement(PieChart_1.default, __assign({}, args, { data: data, palette: chartConstants.PALETTE_MONOCHROME_4_5 })),
        react_1.default.createElement(PieChart_1.default, __assign({}, args, { data: data, palette: chartConstants.PALETTE_MONOCHROME_5_5 }))));
};
exports.Basic = Basic;
/* Basic Donuts */
var BasicDonuts = function (args) {
    /* eslint-disable comma-spacing */
    var data = [
        { x: 'Leslie', y: 60 },
        { x: 'Ron', y: 40 },
        { x: 'Tom', y: 30 },
        { x: 'Gary', y: 20 },
        { x: 'Ben', y: 15 },
    ];
    var style = {
        paddingTop: '4rem',
    };
    return (react_1.default.createElement("section", { style: style },
        react_1.default.createElement(PieChart_1.default, __assign({}, args, { data: data, isDonut: true })),
        react_1.default.createElement(PieChart_1.default, __assign({}, args, { data: data, palette: chartConstants.PALETTE_MONOCHROME_0_5, isDonut: true })),
        react_1.default.createElement(PieChart_1.default, __assign({}, args, { data: data, palette: chartConstants.PALETTE_MONOCHROME_1_5, isDonut: true })),
        react_1.default.createElement(PieChart_1.default, __assign({}, args, { data: data, palette: chartConstants.PALETTE_MONOCHROME_2_5, isDonut: true })),
        react_1.default.createElement(PieChart_1.default, __assign({}, args, { data: data, palette: chartConstants.PALETTE_MONOCHROME_3_5, isDonut: true })),
        react_1.default.createElement(PieChart_1.default, __assign({}, args, { data: data, palette: chartConstants.PALETTE_MONOCHROME_4_5, isDonut: true })),
        react_1.default.createElement(PieChart_1.default, __assign({}, args, { data: data, palette: chartConstants.PALETTE_MONOCHROME_5_5, isDonut: true }))));
};
exports.BasicDonuts = BasicDonuts;
/* Legend */
var LegendTest = function (args) {
    var data = [
        { x: 'Leslie', y: 80 },
        { x: 'Tom', y: 20 },
        { x: 'Ron', y: 10 },
        { x: 'Ann', y: 30 },
    ];
    var palette = chartConstants.PALETTE_7;
    var style = {
        display: 'flex',
        alignItems: 'center',
        paddingTop: '4rem',
    };
    return (react_1.default.createElement("section", { style: style },
        react_1.default.createElement(PieChart_1.default, __assign({}, args, { data: data, palette: palette })),
        react_1.default.createElement(Legend_1.default, null, lodash_1.default.map(data, function (d, index) { return (react_1.default.createElement(Legend_1.default.Item, { key: index, color: palette[index % palette.length], hasPoint: true, pointKind: 1 }, d.x)); }))));
};
exports.LegendTest = LegendTest;
/* Color Map */
var ColorMap = function (args) {
    /* eslint-disable comma-spacing */
    var data = [
        { x: 'Leslie', y: 100 },
        { x: 'Tom', y: 20 },
        { x: 'Ron', y: 10 },
        { x: 'Ann', y: 30 },
        { x: 'Tammy', y: 40 },
    ];
    var style = {
        paddingTop: '4rem',
    };
    return (react_1.default.createElement("section", { style: style },
        react_1.default.createElement(PieChart_1.default, __assign({}, args, { data: data, colorMap: {
                Tammy: chartConstants.COLOR_BAD,
                Leslie: chartConstants.COLOR_GOOD,
            } }))));
};
exports.ColorMap = ColorMap;
/* Percents */
var Percents = function (args) {
    /* eslint-disable comma-spacing */
    var data = [
        { x: 'Leslie', y: 60 },
        { x: 'Ron', y: 40 },
        { x: 'Tom', y: 30 },
        { x: 'Gary', y: 20 },
        { x: 'Ben', y: 15 },
    ];
    var total = lodash_1.default.sum(lodash_1.default.map(data, 'y'));
    var style = {
        paddingTop: '4rem',
    };
    return (react_1.default.createElement("section", { style: style },
        react_1.default.createElement(PieChart_1.default, __assign({}, args, { data: data, yAxisFormatter: function (value) {
                return "".concat(((value / total) * 100).toFixed(1), "%");
            } }))));
};
exports.Percents = Percents;
/* Small With No Stroke Or Hover */
var SmallWithNoStrokeOrHover = function (args) {
    var data = [
        { x: 'Leslie', y: 60 },
        { x: 'Ron', y: 40 },
        { x: 'Tom', y: 30 },
        { x: 'Gary', y: 20 },
        { x: 'Ben', y: 15 },
    ];
    var style = {
        paddingTop: '4rem',
    };
    return (react_1.default.createElement("section", { style: style },
        react_1.default.createElement(PieChart_1.default, __assign({}, args, { margin: {
                top: 0,
                right: 0,
                bottom: 0,
                left: 0,
            }, width: 25, height: 25, data: data, hasStroke: false, isHovering: false }))));
};
exports.SmallWithNoStrokeOrHover = SmallWithNoStrokeOrHover;
